{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e400aebf-e46f-45e8-894b-adf9dd58d138",
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import logging\n",
    "import re\n",
    "import sys\n",
    "import warnings\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "import rpy2.robjects as ro\n",
    "import anndata2ri\n",
    "import rpy2.rinterface_lib.callbacks as rcb\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "from rpy2.rinterface_lib.embedded import RRuntimeError\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as fm\n",
    "plt.rcParams[\"font.family\"] = \"Arial\" if \"Arial\" in [f.name for f in fm.fontManager.ttflist] else \"sans-serif\"\n",
    "plt.rcParams['font.weight'] = 'normal'\n",
    "\n",
    "sys.path.insert(1, \"/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark\")\n",
    "from cellseg_benchmark.adata_utils import plot_pseudobulk_pca\n",
    "from cellseg_benchmark._constants import cell_type_colors\n",
    "from cellseg_benchmark.dea_utils import (\n",
    "    pseudobulk_aggregate_and_filter,\n",
    "    prepare_adata_for_rpy2,\n",
    "    add_group_sample_counts,\n",
    "    add_ensembl_id,\n",
    "    safe_sheet,\n",
    ")\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", message=\".*Observation names are not unique*\")\n",
    "\n",
    "def get_args(test_args=None):  # noqa: D103\n",
    "    p = argparse.ArgumentParser(description=\"DREAM pseudobulk-based DEA\")\n",
    "    p.add_argument(\"cohort\", help=\"Cohort name, e.g. foxf2\")\n",
    "    p.add_argument(\"seg_method\", help=\"Segmentation method, e.g. Cellpose_1_nuclei_model\")\n",
    "    p.add_argument(\"--sample_key\", default=\"sample\", help=\"obs column for donor/sample ID\")\n",
    "    p.add_argument(\"--subset_key\", default=\"cell_type\",\n",
    "                        help=\"obs column used to subset data (e.g. cell_type, cluster, region)\")\n",
    "    p.add_argument(\"--subset_values\", nargs=\"+\", default=None, \n",
    "                        help=\"Values of subset_key to process (default: all unique values)\")\n",
    "    p.add_argument(\"--condition_key\", default=\"genotype\", help=\"obs column for condition (e.g. genotype)\")\n",
    "    p.add_argument(\"--batch_key\", default=\"slide\", \n",
    "                       help=\"optional batch key for inclusion as covariate (default: slide)\")\n",
    "    p.add_argument(\"--ref\", default=\"WT\", help=\"Reference group (default: WT)\")\n",
    "    p.add_argument(\"--test_groups\", nargs=\"+\", default=None,\n",
    "                        help=\"Groups to test vs reference (default: all groups in condition_key except --ref)\")\n",
    "    \n",
    "    p.add_argument(\"--threads\", default=4, help=\"number of threads for DREAM parallelization\")\n",
    "    p.add_argument(\"--min_cells\", type=int, default=15, help=\"Minimum cells per donor\")\n",
    "    p.add_argument(\"--replicates_per_patient\", type=int, default=1,\n",
    "                        help=\"Number of pseudoreplicates per donor\")\n",
    "    p.add_argument(\n",
    "        \"--overwrite\", \n",
    "        type=lambda x: str(x).lower() in [\"true\",\"1\",\"yes\"],\n",
    "        default=True,\n",
    "        help=\"Overwrite existing result files (default: True)\"\n",
    "    )\n",
    "    if test_args is not None:\n",
    "        return p.parse_args(test_args)\n",
    "    else:\n",
    "        return p.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7e7fcf16-9283-478e-b060-8c86ebb8f4d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simulate CLI arguments inside Jupyter\n",
    "args = get_args([\n",
    "    \"foxf2\",\n",
    "    \"Cellpose_1_nuclei_model\",\n",
    "    \"--sample_key\", \"sample\",\n",
    "    \"--subset_key\", \"cell_type\",\n",
    "    \"--condition_key\", \"genotype\",\n",
    "    \"--ref\", \"WT\"\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7c6c103f-4b80-49a4-b919-3c6af1d00794",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simulate CLI arguments inside Jupyter\n",
    "args = get_args([\n",
    "    \"aging\",\n",
    "    \"Cellpose_1_nuclei_model\",\n",
    "    \"--sample_key\", \"sample\",\n",
    "    \"--subset_key\", \"cell_type\",\n",
    "    \"--condition_key\", \"condition\",\n",
    "    \"--ref\", \"WT_3\",\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "daae90b4-c06d-420d-b378-d6a7be7e5849",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Namespace(cohort='aging', seg_method='Cellpose_1_nuclei_model', sample_key='sample', subset_key='cell_type', subset_values=None, condition_key='condition', batch_key='slide', ref='WT_3', test_groups=None, threads=4, min_cells=15, replicates_per_patient=1, overwrite=True)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8edfe4fd-09e5-4233-94d2-bbafda710f07",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Logger setup\n",
    "logger = logging.getLogger(\"dea\")\n",
    "if not logger.handlers:\n",
    "    handler = logging.StreamHandler()\n",
    "    handler.setFormatter(logging.Formatter(\"%(asctime)s [%(levelname)s]: %(message)s\"))\n",
    "    logger.addHandler(handler)\n",
    "    logger.setLevel(logging.INFO)\n",
    "    logger.propagate = False\n",
    "rcb.logger.handlers = logger.handlers\n",
    "rcb.consolewrite_print   = lambda x: logger.debug(f\"R: {x.strip()}\")\n",
    "rcb.consolewrite_error   = lambda x: (_ for _ in ()).throw(RRuntimeError(x.strip()))\n",
    "rcb.consolewrite_message = lambda x: logger.info(f\"R: {x.strip()}\")\n",
    "rcb.consolewrite_warn = lambda x: (logger.warning if x.lstrip().lower().startswith(\"warning\") else logger.info)(f\"R: {x.strip()}\")\n",
    "setattr(rcb, \"consolewrite_warnerror\", rcb.consolewrite_warn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d02203c5-2be9-4397-bfb0-5e840ec368ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# R setup\n",
    "conv = ro.default_converter + ro.pandas2ri.converter + anndata2ri.converter\n",
    "r_script = Path(sys.path[1]) / \"cellseg_benchmark\" / \"dea_utils.r\"\n",
    "ro.r[\"source\"](str(r_script))\n",
    "edgeR_loop = ro.globalenv[\"edgeR_loop\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "da1cc684-06b6-4f10-96f6-1444b9f2f9aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = Path(\"/dss/dssfs03/pn52re/pn52re-dss-0001/cellseg-benchmark\")\n",
    "method_path = base_path / \"analysis\" / args.cohort / args.seg_method\n",
    "output_dir = method_path / \"dea\"\n",
    "output_dir.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "20911229-e913-4688-927d-d2680eb92920",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-29 16:43:58,347 [INFO]: Loading integrated AnnData...\n"
     ]
    }
   ],
   "source": [
    "logger.info(\"Loading integrated AnnData...\")\n",
    "adata = sc.read_h5ad(method_path / \"adatas\" / \"adata_integrated.h5ad.gz\")\n",
    "adata.obs[\"cell_type\"] = adata.obs[\"cell_type_revised\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1b7e85a-ec8b-4a76-aaf7-27fef4575e60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# re-group cell types\n",
    "adata.obs[\"cell_type\"] = (\n",
    "    adata.obs[\"cell_type\"]\n",
    "    .astype(str)\n",
    "    .replace({\n",
    "        \"Tanycytes\": \"Ependymal\",\n",
    "        \"Astroependymal\": \"Astrocytes\",\n",
    "        \"Neurons-Glyc-Gaba\" : \"Neurons-Other\"\n",
    "    })\n",
    "    .astype(\"category\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "961b6419-bba0-441d-973d-a05d840c8b1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell_type\n",
       "Oligodendrocytes            131904\n",
       "Neurons-Glut                124130\n",
       "Neurons-Dopa                 77224\n",
       "Astrocytes                   75424\n",
       "ECs                          49431\n",
       "Neurons-Granule-Immature     42337\n",
       "Neurons-Other                40619\n",
       "Undefined                    32012\n",
       "Microglia                    29106\n",
       "OPCs                         20832\n",
       "Pericytes                    15056\n",
       "VLMCs                         8346\n",
       "Ependymal                     6539\n",
       "SMCs                          5928\n",
       "Choroid-Plexus                5657\n",
       "Neurons-Gaba                  3499\n",
       "BAMs                          1933\n",
       "Immune-Other                   566\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"cell_type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c6f7b45c-ac94-4a48-ac90-f69943db7e16",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-29 16:51:29,626 [INFO]: Test groups inferred from 'condition': WT_6, WT_12, WT_18, WT_24\n",
      "2025-10-29 16:51:30,783 [INFO]: WT_3: 3 samples → aging_s8_r1, aging_s10_r1, aging_s10_r2\n",
      "2025-10-29 16:51:30,811 [INFO]: WT_6: 3 samples → aging_s1_r0, aging_s5_r1, aging_s7_r2\n",
      "2025-10-29 16:51:30,819 [INFO]: WT_12: 3 samples → aging_s5_r2, aging_s8_r0, aging_s12_r0\n",
      "2025-10-29 16:51:30,831 [INFO]: WT_18: 3 samples → aging_s6_r0, aging_s8_r2, aging_s11_r0\n",
      "2025-10-29 16:51:30,847 [INFO]: WT_24: 3 samples → aging_s10_r0, aging_s11_r1, aging_s11_r2\n"
     ]
    }
   ],
   "source": [
    "# Clean up group names for R conversion\n",
    "adata.obs[args.subset_key] = [\n",
    "    key.replace(\" \", \"_\").replace(\"-\", \"_\").replace(\"/\", \"_\")\n",
    "      .replace(\"+\", \"\").replace(\"(\", \"\").replace(\")\", \"\") \n",
    "    if not pd.isna(key) else key \n",
    "    for key in adata.obs[args.subset_key]\n",
    "]\n",
    "for col in [args.condition_key, args.sample_key, args.subset_key]:\n",
    "    adata.obs[col] = adata.obs[col].astype(\"category\")\n",
    "if args.test_groups is None:\n",
    "    groups = adata.obs[args.condition_key].dropna().unique().tolist()\n",
    "    args.test_groups = [g for g in groups if g != args.ref]\n",
    "    msg = (f\"Test groups inferred from '{args.condition_key}': {', '.join(map(str, args.test_groups))}\"\n",
    "           if args.test_groups else\n",
    "           f\"No test groups found different from ref '{args.ref}' under '{args.condition_key}'.\")\n",
    "    logger.info(msg)\n",
    "    \n",
    "# print sample overview\n",
    "if args.subset_values is None:\n",
    "    groups_to_process = sorted(adata.obs[args.subset_key].unique())\n",
    "else:\n",
    "    groups_to_process = args.subset_values\n",
    "for cond, df in adata.obs.groupby(args.condition_key, observed=True):\n",
    "    samples = df[args.sample_key].unique()\n",
    "    logger.info(f\"{cond}: {len(samples)} samples → {', '.join(samples)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ef2cc40c-5823-40d1-ac80-ecbdbbcdf3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['n_cells_sum'] = adata.obs.groupby([args.subset_key, args.sample_key], observed=True)[args.sample_key].transform('count')\n",
    "adata.obs['volume_mean'] = adata.obs.groupby([args.subset_key, args.sample_key], observed=True)['volume_final'].transform('mean')\n",
    "adata.obs['volume_sum'] = adata.obs.groupby([args.subset_key, args.sample_key], observed=True)['volume_final'].transform('sum')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ddbcd813-71a1-4393-8853-9959db7659b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers['counts'].copy()\n",
    "assert np.issubdtype(adata.X.dtype, np.integer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6b6290b9-4885-4d84-b97c-32cd390963cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_to_keep = [\n",
    "    args.condition_key,\n",
    "    args.subset_key,\n",
    "    args.sample_key,\n",
    "    \"n_cells_sum\",\n",
    "    \"volume_mean\",\n",
    "    \"volume_sum\",\n",
    "]\n",
    "if args.batch_key and args.batch_key in adata.obs.columns:\n",
    "    obs_to_keep.append(args.batch_key)\n",
    "else:\n",
    "    logger.warning(\n",
    "        f\"Batch column '{args.batch_key}' was not found in adata.obs \"\n",
    "        \"or is invalid. Continuing **without** a batch covariate.\"\n",
    "    )\n",
    "    args.batch_key = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6976c500-339e-48a1-96d2-e0b8eaaa9996",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-29 16:51:35,886 [INFO]: Run pseudobulking...\n",
      "2025-10-29 16:51:35,902 [INFO]: Processing Astrocytes (1/18)\n",
      "2025-10-29 16:51:46,618 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:52:10,807 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:52:12,311 [INFO]: Processing BAMs (2/18)\n",
      "2025-10-29 16:52:15,360 [INFO]: \tProcessing donor 1/14...\n",
      "2025-10-29 16:52:30,511 [INFO]: \tProcessing donor 14/14...\n",
      "2025-10-29 16:52:31,752 [INFO]: Processing Choroid_Plexus (3/18)\n",
      "2025-10-29 16:52:34,537 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:52:51,545 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:52:53,507 [INFO]: Processing ECs (4/18)\n",
      "2025-10-29 16:53:04,601 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:53:31,329 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:53:34,914 [INFO]: Processing Ependymal (5/18)\n",
      "2025-10-29 16:53:39,427 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:53:56,973 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:53:57,937 [INFO]: Processing Immune_Other (6/18)\n",
      "2025-10-29 16:53:59,573 [INFO]: \tDropping samples with <15 cells: {'aging_s10_r2': 12}\n",
      "2025-10-29 16:53:59,575 [INFO]: \tProcessing donor 1/5...\n",
      "2025-10-29 16:54:03,702 [INFO]: \tProcessing donor 5/5...\n",
      "2025-10-29 16:54:05,137 [INFO]: Processing Microglia (7/18)\n",
      "2025-10-29 16:54:16,874 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:54:41,018 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:54:42,398 [INFO]: Processing Neurons_Dopa (8/18)\n",
      "2025-10-29 16:54:56,765 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:55:17,476 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:55:18,553 [INFO]: Processing Neurons_Gaba (9/18)\n",
      "2025-10-29 16:55:20,219 [INFO]: \tProcessing donor 1/14...\n",
      "2025-10-29 16:55:30,361 [INFO]: \tProcessing donor 14/14...\n",
      "2025-10-29 16:55:31,169 [INFO]: Processing Neurons_Glut (10/18)\n",
      "2025-10-29 16:55:44,263 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:56:09,247 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:56:11,559 [INFO]: Processing Neurons_Granule_Immature (11/18)\n",
      "2025-10-29 16:56:19,658 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:56:50,594 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:56:52,634 [INFO]: Processing Neurons_Other (12/18)\n",
      "2025-10-29 16:57:00,489 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:57:21,113 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:57:22,622 [INFO]: Processing OPCs (13/18)\n",
      "2025-10-29 16:57:26,054 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:57:39,084 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:57:40,021 [INFO]: Processing Oligodendrocytes (14/18)\n",
      "2025-10-29 16:57:58,374 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:58:35,876 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:58:38,828 [INFO]: Processing Pericytes (15/18)\n",
      "2025-10-29 16:58:43,520 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:58:59,923 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:59:00,839 [INFO]: Processing SMCs (16/18)\n",
      "2025-10-29 16:59:03,392 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:59:13,755 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:59:14,520 [INFO]: Processing Undefined (17/18)\n",
      "2025-10-29 16:59:18,298 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:59:35,029 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 16:59:36,486 [INFO]: Processing VLMCs (18/18)\n",
      "2025-10-29 16:59:41,051 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 16:59:59,137 [INFO]: \tProcessing donor 15/15...\n",
      "2025-10-29 17:00:02,243 [INFO]: Processing all data (no subsetting by cell_type)\n",
      "2025-10-29 17:00:45,622 [INFO]: \tProcessing donor 1/15...\n",
      "2025-10-29 17:02:58,990 [INFO]: \tProcessing donor 15/15...\n"
     ]
    }
   ],
   "source": [
    "logger.info(f\"Run pseudobulking...\")\n",
    "total = len(groups_to_process)\n",
    "logger.info(f\"Processing {groups_to_process[0]} ({1}/{total})\")\n",
    "adata_pb = pseudobulk_aggregate_and_filter(\n",
    "    adata,\n",
    "    subset_value=groups_to_process[0],\n",
    "    sample_key=args.sample_key,\n",
    "    subset_key=args.subset_key,\n",
    "    obs_to_keep=obs_to_keep,\n",
    "    min_cells=args.min_cells,\n",
    "    replicates_per_patient=args.replicates_per_patient,\n",
    "    logger=logger\n",
    ")\n",
    "for i, group in enumerate(groups_to_process[1:], 2):\n",
    "    logger.info(f\"Processing {group} ({i}/{total})\")\n",
    "    adata_pb_i = pseudobulk_aggregate_and_filter(\n",
    "        adata,\n",
    "        subset_value=group,\n",
    "        sample_key=args.sample_key,\n",
    "        subset_key=args.subset_key,\n",
    "        obs_to_keep=obs_to_keep,\n",
    "        min_cells=args.min_cells,\n",
    "        replicates_per_patient=args.replicates_per_patient,\n",
    "        logger=logger\n",
    "    )\n",
    "    adata_pb = ad.concat([adata_pb, adata_pb_i], join=\"outer\", label=None, index_unique=None)\n",
    "adata_pb.obs_names_make_unique()\n",
    "# Add whole brain: pseudobulk per sample across all groups\n",
    "adata_pb_i = pseudobulk_aggregate_and_filter(\n",
    "    adata,\n",
    "    subset_value=None,\n",
    "    sample_key=args.sample_key,\n",
    "    subset_key=args.subset_key,\n",
    "    obs_to_keep=obs_to_keep,\n",
    "    min_cells=args.min_cells,\n",
    "    replicates_per_patient=args.replicates_per_patient,\n",
    "    logger=logger\n",
    ")\n",
    "adata_pb_i.obs[args.subset_key] = \"all\"\n",
    "adata_pb = ad.concat([adata_pb, adata_pb_i])\n",
    "del adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d3c04584-12f1-4f6b-a08a-a2cfc75f9b39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-29 17:03:10,730 [INFO]: Skipping PCA plot, file already exists: /dss/dssfs03/pn52re/pn52re-dss-0001/cellseg-benchmark/analysis/aging/Cellpose_1_nuclei_model/dea/pseudobulk_pca.png\n"
     ]
    }
   ],
   "source": [
    "plot_pseudobulk_pca(adata_pb, args, output_dir, cell_type_colors, logger)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0e04c200-d74e-4dc3-8809-c9b7bca337bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-29 17:03:10,747 [INFO]: Run DEA...\n"
     ]
    }
   ],
   "source": [
    "logger.info(f\"Run DEA...\")\n",
    "adata_pb = prepare_adata_for_rpy2(adata_pb, key=args.subset_key)\n",
    "adatas_pb = {}\n",
    "unique_groups = adata_pb.obs[args.subset_key].unique().tolist()\n",
    "for key in unique_groups:\n",
    "    tmp = adata_pb[adata_pb.obs[args.subset_key] == key].copy()\n",
    "    adatas_pb[key] = tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08958278-7c39-401b-9652-165355ef4bd4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cbe6ff8-2ee5-4d98-b5a4-bb4c7a0ad0fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eba7c46-f4a8-4fe5-8e8b-c5afcd770c55",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "18e2c7eb-1e53-495e-9958-1ce1f5d10a79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# explore sce file in r\n",
    "# skip when testing script!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9d8a10b9-f995-46de-8c68-134c49e55aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "75ad24af-6cb1-4ecb-ab63-7faab07970e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GLKO', 'ECKO', 'PCKO']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args.test_groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "720c3e8a-b1f2-4cc6-8cf1-915312198b3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_i = \"ECs\"\n",
    "adata = adatas_pb[group_i]\n",
    "condition_col = args.condition_key\n",
    "test_groups = args.test_groups\n",
    "ref_group = args.ref\n",
    "batch_col = args.batch_key\n",
    "threads = args.threads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0681222c-03a9-488b-9deb-22500d9a33e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'slide'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "batch_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "574b8d05-8b8b-49ec-9e59-0bc5a20d8c5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GLKO', 'ECKO', 'PCKO']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c24088ac-c9c5-4c43-a27a-96243dbdbc13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "threads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "79094baa-2dcd-4251-881b-7a6609bf8524",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>slide</th>\n",
       "      <th>n_cells_sum</th>\n",
       "      <th>volume_mean</th>\n",
       "      <th>volume_sum</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s1_r0_0-4</th>\n",
       "      <td>ECKO</td>\n",
       "      <td>ECs</td>\n",
       "      <td>1</td>\n",
       "      <td>3770</td>\n",
       "      <td>1432.306890</td>\n",
       "      <td>5.399797e+06</td>\n",
       "      <td>foxf2_s1_r0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s1_r1_0-4</th>\n",
       "      <td>WT</td>\n",
       "      <td>ECs</td>\n",
       "      <td>1</td>\n",
       "      <td>3244</td>\n",
       "      <td>1441.239386</td>\n",
       "      <td>4.675381e+06</td>\n",
       "      <td>foxf2_s1_r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s2_r1_0-4</th>\n",
       "      <td>WT</td>\n",
       "      <td>ECs</td>\n",
       "      <td>2</td>\n",
       "      <td>3709</td>\n",
       "      <td>1321.529465</td>\n",
       "      <td>4.901553e+06</td>\n",
       "      <td>foxf2_s2_r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s2_r2_0-4</th>\n",
       "      <td>PCKO</td>\n",
       "      <td>ECs</td>\n",
       "      <td>2</td>\n",
       "      <td>3730</td>\n",
       "      <td>1422.418953</td>\n",
       "      <td>5.305623e+06</td>\n",
       "      <td>foxf2_s2_r2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s4_r0_0-3</th>\n",
       "      <td>PCKO</td>\n",
       "      <td>ECs</td>\n",
       "      <td>4</td>\n",
       "      <td>4136</td>\n",
       "      <td>1372.144486</td>\n",
       "      <td>5.675190e+06</td>\n",
       "      <td>foxf2_s4_r0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s4_r1_0-4</th>\n",
       "      <td>ECKO</td>\n",
       "      <td>ECs</td>\n",
       "      <td>4</td>\n",
       "      <td>2965</td>\n",
       "      <td>1478.098312</td>\n",
       "      <td>4.382561e+06</td>\n",
       "      <td>foxf2_s4_r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s5_r0_0-4</th>\n",
       "      <td>PCKO</td>\n",
       "      <td>ECs</td>\n",
       "      <td>5</td>\n",
       "      <td>4889</td>\n",
       "      <td>1326.376653</td>\n",
       "      <td>6.484655e+06</td>\n",
       "      <td>foxf2_s5_r0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s5_r1_0-4</th>\n",
       "      <td>ECKO</td>\n",
       "      <td>ECs</td>\n",
       "      <td>5</td>\n",
       "      <td>3511</td>\n",
       "      <td>1172.656272</td>\n",
       "      <td>4.117196e+06</td>\n",
       "      <td>foxf2_s5_r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s6_r0_0-4</th>\n",
       "      <td>GLKO</td>\n",
       "      <td>ECs</td>\n",
       "      <td>6</td>\n",
       "      <td>2986</td>\n",
       "      <td>1495.124579</td>\n",
       "      <td>4.464442e+06</td>\n",
       "      <td>foxf2_s6_r0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s6_r1_0-4</th>\n",
       "      <td>GLKO</td>\n",
       "      <td>ECs</td>\n",
       "      <td>6</td>\n",
       "      <td>3486</td>\n",
       "      <td>1512.162401</td>\n",
       "      <td>5.271398e+06</td>\n",
       "      <td>foxf2_s6_r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s6_r2_0-4</th>\n",
       "      <td>WT</td>\n",
       "      <td>ECs</td>\n",
       "      <td>6</td>\n",
       "      <td>3646</td>\n",
       "      <td>1502.238159</td>\n",
       "      <td>5.477160e+06</td>\n",
       "      <td>foxf2_s6_r2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s7_r0_0-4</th>\n",
       "      <td>WT</td>\n",
       "      <td>ECs</td>\n",
       "      <td>7</td>\n",
       "      <td>3441</td>\n",
       "      <td>1520.296814</td>\n",
       "      <td>5.231341e+06</td>\n",
       "      <td>foxf2_s7_r0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s7_r1_0-4</th>\n",
       "      <td>GLKO</td>\n",
       "      <td>ECs</td>\n",
       "      <td>7</td>\n",
       "      <td>3323</td>\n",
       "      <td>1443.922011</td>\n",
       "      <td>4.798153e+06</td>\n",
       "      <td>foxf2_s7_r1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      genotype cell_type slide  n_cells_sum  volume_mean  \\\n",
       "donor_foxf2_s1_r0_0-4     ECKO       ECs     1         3770  1432.306890   \n",
       "donor_foxf2_s1_r1_0-4       WT       ECs     1         3244  1441.239386   \n",
       "donor_foxf2_s2_r1_0-4       WT       ECs     2         3709  1321.529465   \n",
       "donor_foxf2_s2_r2_0-4     PCKO       ECs     2         3730  1422.418953   \n",
       "donor_foxf2_s4_r0_0-3     PCKO       ECs     4         4136  1372.144486   \n",
       "donor_foxf2_s4_r1_0-4     ECKO       ECs     4         2965  1478.098312   \n",
       "donor_foxf2_s5_r0_0-4     PCKO       ECs     5         4889  1326.376653   \n",
       "donor_foxf2_s5_r1_0-4     ECKO       ECs     5         3511  1172.656272   \n",
       "donor_foxf2_s6_r0_0-4     GLKO       ECs     6         2986  1495.124579   \n",
       "donor_foxf2_s6_r1_0-4     GLKO       ECs     6         3486  1512.162401   \n",
       "donor_foxf2_s6_r2_0-4       WT       ECs     6         3646  1502.238159   \n",
       "donor_foxf2_s7_r0_0-4       WT       ECs     7         3441  1520.296814   \n",
       "donor_foxf2_s7_r1_0-4     GLKO       ECs     7         3323  1443.922011   \n",
       "\n",
       "                         volume_sum       sample  \n",
       "donor_foxf2_s1_r0_0-4  5.399797e+06  foxf2_s1_r0  \n",
       "donor_foxf2_s1_r1_0-4  4.675381e+06  foxf2_s1_r1  \n",
       "donor_foxf2_s2_r1_0-4  4.901553e+06  foxf2_s2_r1  \n",
       "donor_foxf2_s2_r2_0-4  5.305623e+06  foxf2_s2_r2  \n",
       "donor_foxf2_s4_r0_0-3  5.675190e+06  foxf2_s4_r0  \n",
       "donor_foxf2_s4_r1_0-4  4.382561e+06  foxf2_s4_r1  \n",
       "donor_foxf2_s5_r0_0-4  6.484655e+06  foxf2_s5_r0  \n",
       "donor_foxf2_s5_r1_0-4  4.117196e+06  foxf2_s5_r1  \n",
       "donor_foxf2_s6_r0_0-4  4.464442e+06  foxf2_s6_r0  \n",
       "donor_foxf2_s6_r1_0-4  5.271398e+06  foxf2_s6_r1  \n",
       "donor_foxf2_s6_r2_0-4  5.477160e+06  foxf2_s6_r2  \n",
       "donor_foxf2_s7_r0_0-4  5.231341e+06  foxf2_s7_r0  \n",
       "donor_foxf2_s7_r1_0-4  4.798153e+06  foxf2_s7_r1  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5b8ff1ad-c083-4120-b838-1c0046763a75",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Looking for: ['bioconda::bioconductor-variancepartition']\n",
      "\n",
      "conda-forge/linux-64                                        Using cache\n",
      "conda-forge/noarch                                          Using cache\n",
      "bioconda/linux-64                                           Using cache\n",
      "bioconda/noarch                                             Using cache\n",
      "\n",
      "Pinned packages:\n",
      "  - python 3.12.*\n",
      "\n",
      "\n",
      "Transaction\n",
      "\n",
      "  Prefix: /home/ubuntu/miniforge3\n",
      "\n",
      "  Updating specs:\n",
      "\n",
      "   - bioconda::bioconductor-variancepartition\n",
      "   - ca-certificates\n",
      "   - certifi\n",
      "   - openssl\n",
      "\n",
      "\n",
      "  Package                           Version  Build          Channel        Size\n",
      "─────────────────────────────────────────────────────────────────────────────────\n",
      "  Reinstall:\n",
      "─────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  \u001b[32mo bioconductor-variancepartition\u001b[0m   1.36.2  r44hdfd78af_0  bioconda\u001b[32m     Cached\u001b[0m\n",
      "\n",
      "  Summary:\n",
      "\n",
      "  Reinstall: 1 packages\n",
      "\n",
      "  Total download: 0 B\n",
      "\n",
      "─────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "\n",
      "\u001b[?25l\u001b[2K\u001b[0G\u001b[?25h"
     ]
    }
   ],
   "source": [
    "#! mamba install bioconda::bioconductor-variancepartition -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e265bd52-d4d4-416c-a29c-1a47e330ed11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "data.table 1.17.8 using 2 threads (see ?getDTthreads).  Latest news: r-datatable.com\n",
       "\n",
       "Attaching package: ‘data.table’\n",
       "\n",
       "The following object is masked from ‘package:SummarizedExperiment’:\n",
       "\n",
       "    shift\n",
       "\n",
       "The following object is masked from ‘package:GenomicRanges’:\n",
       "\n",
       "    shift\n",
       "\n",
       "The following object is masked from ‘package:IRanges’:\n",
       "\n",
       "    shift\n",
       "\n",
       "The following objects are masked from ‘package:S4Vectors’:\n",
       "\n",
       "    first, second\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "library(\"variancePartition\")\n",
    "library(\"edgeR\")\n",
    "library(\"BiocParallel\")\n",
    "library(data.table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d446d8bf-21a0-4d8a-ab7b-d28812ad425d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "49d49da3-ceba-4127-b9ab-ec82109310fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'genotype'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e15c3603-0501-43ee-b9ec-6c4ec3dd2a55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ECs: GLKO/ECKO/PCKO vs WT\n",
       "Design: ~ genotype + slide\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -c conv -i adata -i condition_col -i test_groups -i group_i -i ref_group -i batch_col -i threads -o design\n",
    "\n",
    "condition <- droplevels(factor(colData(adata)[[condition_col]]))\n",
    "if (is.null(condition) || nlevels(condition) < 2) {\n",
    "    message(group_i, \": skip (condition invalid or <2 levels)\"); return(NULL)\n",
    "}\n",
    "\n",
    "present <- intersect(test_groups, levels(condition))\n",
    "if (!(ref_group %in% levels(condition)) || !length(present)) {\n",
    "    message(group_i, \": skip (ref_group/test_groups missing)\"); return(NULL)\n",
    "}\n",
    "\n",
    "missing <- setdiff(test_groups, present)\n",
    "message(group_i, \": \", paste(present, collapse=\"/\"), \" vs \", ref_group,\n",
    "        if (length(missing)) paste0(\" (missing:\", paste(missing, collapse=\",\"), \")\") else \"\")\n",
    "\n",
    "param <- BiocParallel::SnowParam(as.numeric(threads), \"SOCK\", progressbar = T)\n",
    "\n",
    "\n",
    "\n",
    "df <- as.data.frame(colData(adata))\n",
    "\n",
    "condition <- droplevels(factor(colData(adata)[[condition_col]]))\n",
    "if (!(ref_group %in% levels(condition))) stop(\"ref_group not found in condition levels.\")\n",
    "condition <- stats::relevel(condition, ref = ref_group)\n",
    "if (levels(condition)[1L] != ref_group) stop(\"ref_group is not the first level after relevel().\")\n",
    "df$condition <- condition\n",
    "\n",
    "# use batch only if it exists AND has >1 level\n",
    "use_batch <- !is.null(batch_col) &&\n",
    "             batch_col %in% colnames(colData(adata)) &&\n",
    "             nlevels(droplevels(factor(colData(adata)[[batch_col]]))) > 1\n",
    "\n",
    "if (use_batch) {\n",
    "    df$batch <- droplevels(factor(df[[batch_col]]))\n",
    "    form <- ~ condition + batch\n",
    "    message(\"Design: ~ \", condition_col, \" + \", batch_col)\n",
    "} else {\n",
    "    form <- ~ condition\n",
    "    message(\"Design: ~ \", condition_col, if (!is.null(batch_col)) \" (batch ignored: <2 levels or missing)\" else \"\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "45c7043b-925e-4697-82f8-5e485054fb04",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%R -o libsize -o vol_sum\n",
    "#libsize <- colSums(SummarizedExperiment::assay(adata, \"X\"))\n",
    "#vol_sum <- colData(adata)$volume_sum\n",
    "#cor(libsize, vol_sum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8efed266-dd9b-49fd-afe3-8b4eb82a7d6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o s -o c\n",
    "min_count = 2\n",
    "y <- edgeR::DGEList(assay(adata, \"X\"), group = df$condition)\n",
    "\n",
    "# replace library sizes with volume\n",
    "vol <- colData(adata)[[\"volume_sum\"]]\n",
    "if (is.null(vol) || any(vol <= 0)) stop(\"Invalid volume in colData(sce), expects 'volume_sum'.\")\n",
    "y$samples$lib.size <- vol\n",
    "y <- edgeR::calcNormFactors(y)\n",
    "\n",
    "s <- y$samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ece4bf4d-cbee-48ff-abf6-05f1c1224376",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>lib.size</th>\n",
       "      <th>norm.factors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s1_r0_0-4</th>\n",
       "      <td>ECKO</td>\n",
       "      <td>5.399797e+06</td>\n",
       "      <td>1.326654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s1_r1_0-4</th>\n",
       "      <td>WT</td>\n",
       "      <td>4.675381e+06</td>\n",
       "      <td>1.333670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s2_r1_0-4</th>\n",
       "      <td>WT</td>\n",
       "      <td>4.901553e+06</td>\n",
       "      <td>0.871858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s2_r2_0-4</th>\n",
       "      <td>PCKO</td>\n",
       "      <td>5.305623e+06</td>\n",
       "      <td>0.914129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s4_r0_0-3</th>\n",
       "      <td>PCKO</td>\n",
       "      <td>5.675190e+06</td>\n",
       "      <td>1.290681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s4_r1_0-4</th>\n",
       "      <td>ECKO</td>\n",
       "      <td>4.382561e+06</td>\n",
       "      <td>1.403576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s5_r0_0-4</th>\n",
       "      <td>PCKO</td>\n",
       "      <td>6.484655e+06</td>\n",
       "      <td>0.785730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s5_r1_0-4</th>\n",
       "      <td>ECKO</td>\n",
       "      <td>4.117196e+06</td>\n",
       "      <td>0.596620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s6_r0_0-4</th>\n",
       "      <td>GLKO</td>\n",
       "      <td>4.464442e+06</td>\n",
       "      <td>0.975105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s6_r1_0-4</th>\n",
       "      <td>GLKO</td>\n",
       "      <td>5.271398e+06</td>\n",
       "      <td>0.851289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s6_r2_0-4</th>\n",
       "      <td>WT</td>\n",
       "      <td>5.477160e+06</td>\n",
       "      <td>0.882613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s7_r0_0-4</th>\n",
       "      <td>WT</td>\n",
       "      <td>5.231341e+06</td>\n",
       "      <td>1.119857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_foxf2_s7_r1_0-4</th>\n",
       "      <td>GLKO</td>\n",
       "      <td>4.798153e+06</td>\n",
       "      <td>1.017779</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      group      lib.size  norm.factors\n",
       "donor_foxf2_s1_r0_0-4  ECKO  5.399797e+06      1.326654\n",
       "donor_foxf2_s1_r1_0-4    WT  4.675381e+06      1.333670\n",
       "donor_foxf2_s2_r1_0-4    WT  4.901553e+06      0.871858\n",
       "donor_foxf2_s2_r2_0-4  PCKO  5.305623e+06      0.914129\n",
       "donor_foxf2_s4_r0_0-3  PCKO  5.675190e+06      1.290681\n",
       "donor_foxf2_s4_r1_0-4  ECKO  4.382561e+06      1.403576\n",
       "donor_foxf2_s5_r0_0-4  PCKO  6.484655e+06      0.785730\n",
       "donor_foxf2_s5_r1_0-4  ECKO  4.117196e+06      0.596620\n",
       "donor_foxf2_s6_r0_0-4  GLKO  4.464442e+06      0.975105\n",
       "donor_foxf2_s6_r1_0-4  GLKO  5.271398e+06      0.851289\n",
       "donor_foxf2_s6_r2_0-4    WT  5.477160e+06      0.882613\n",
       "donor_foxf2_s7_r0_0-4    WT  5.231341e+06      1.119857\n",
       "donor_foxf2_s7_r1_0-4  GLKO  4.798153e+06      1.017779"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a0b5fac1-bec0-4e13-b1a0-56664c84af26",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o dt_all\n",
    "vobjDream <- variancePartition::voomWithDreamWeights(y, form, df, BPPARAM = param)\n",
    "\n",
    "fitmm <- variancePartition::dream(vobjDream, form, df)\n",
    "fitmm <- variancePartition::eBayes(fitmm)\n",
    "\n",
    "res_list <- lapply(test_groups, function(grp) {\n",
    "    tt <- variancePartition::topTable(\n",
    "        fitmm,\n",
    "        coef = paste0(\"condition\", grp),\n",
    "        number = Inf,\n",
    "        sort.by = \"p\"\n",
    "    )\n",
    "    as.data.table(tt, keep.rownames = \"gene\")[, test_group := grp]\n",
    "})\n",
    "dt_all <- rbindlist(res_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ad7374df-71d3-4311-8a12-1d1575367d6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>t</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>B</th>\n",
       "      <th>test_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Foxf2</td>\n",
       "      <td>-2.890146</td>\n",
       "      <td>9.004545</td>\n",
       "      <td>-12.591761</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.001092</td>\n",
       "      <td>5.207381</td>\n",
       "      <td>GLKO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Slc16a1</td>\n",
       "      <td>-0.791860</td>\n",
       "      <td>10.476767</td>\n",
       "      <td>-7.039470</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.033544</td>\n",
       "      <td>1.623069</td>\n",
       "      <td>GLKO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Depp1</td>\n",
       "      <td>0.775995</td>\n",
       "      <td>8.260951</td>\n",
       "      <td>5.792303</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.080380</td>\n",
       "      <td>0.392768</td>\n",
       "      <td>GLKO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Atp13a5</td>\n",
       "      <td>-0.806453</td>\n",
       "      <td>7.167773</td>\n",
       "      <td>-4.311647</td>\n",
       "      <td>0.002856</td>\n",
       "      <td>0.292346</td>\n",
       "      <td>-1.344844</td>\n",
       "      <td>GLKO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Abcg2</td>\n",
       "      <td>0.518439</td>\n",
       "      <td>10.808551</td>\n",
       "      <td>4.293942</td>\n",
       "      <td>0.002923</td>\n",
       "      <td>0.292346</td>\n",
       "      <td>-1.466834</td>\n",
       "      <td>GLKO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1496</th>\n",
       "      <td>Flrt2</td>\n",
       "      <td>-0.002631</td>\n",
       "      <td>8.413505</td>\n",
       "      <td>-0.010844</td>\n",
       "      <td>0.991625</td>\n",
       "      <td>0.999043</td>\n",
       "      <td>-6.056041</td>\n",
       "      <td>PCKO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1497</th>\n",
       "      <td>Icam1</td>\n",
       "      <td>-0.003221</td>\n",
       "      <td>7.277082</td>\n",
       "      <td>-0.009001</td>\n",
       "      <td>0.993049</td>\n",
       "      <td>0.999043</td>\n",
       "      <td>-5.965048</td>\n",
       "      <td>PCKO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1498</th>\n",
       "      <td>Rad54b</td>\n",
       "      <td>-0.000727</td>\n",
       "      <td>7.183502</td>\n",
       "      <td>-0.004880</td>\n",
       "      <td>0.996231</td>\n",
       "      <td>0.999294</td>\n",
       "      <td>-5.985783</td>\n",
       "      <td>PCKO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1499</th>\n",
       "      <td>Spp1</td>\n",
       "      <td>-0.000469</td>\n",
       "      <td>6.090952</td>\n",
       "      <td>-0.001511</td>\n",
       "      <td>0.998833</td>\n",
       "      <td>0.999294</td>\n",
       "      <td>-5.880505</td>\n",
       "      <td>PCKO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1500</th>\n",
       "      <td>Slc3a2</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>10.460773</td>\n",
       "      <td>0.000914</td>\n",
       "      <td>0.999294</td>\n",
       "      <td>0.999294</td>\n",
       "      <td>-6.133592</td>\n",
       "      <td>PCKO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1500 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene     logFC    AveExpr          t   P.Value  adj.P.Val         B  \\\n",
       "1       Foxf2 -2.890146   9.004545 -12.591761  0.000002   0.001092  5.207381   \n",
       "2     Slc16a1 -0.791860  10.476767  -7.039470  0.000134   0.033544  1.623069   \n",
       "3       Depp1  0.775995   8.260951   5.792303  0.000482   0.080380  0.392768   \n",
       "4     Atp13a5 -0.806453   7.167773  -4.311647  0.002856   0.292346 -1.344844   \n",
       "5       Abcg2  0.518439  10.808551   4.293942  0.002923   0.292346 -1.466834   \n",
       "...       ...       ...        ...        ...       ...        ...       ...   \n",
       "1496    Flrt2 -0.002631   8.413505  -0.010844  0.991625   0.999043 -6.056041   \n",
       "1497    Icam1 -0.003221   7.277082  -0.009001  0.993049   0.999043 -5.965048   \n",
       "1498   Rad54b -0.000727   7.183502  -0.004880  0.996231   0.999294 -5.985783   \n",
       "1499     Spp1 -0.000469   6.090952  -0.001511  0.998833   0.999294 -5.880505   \n",
       "1500   Slc3a2  0.000141  10.460773   0.000914  0.999294   0.999294 -6.133592   \n",
       "\n",
       "     test_group  \n",
       "1          GLKO  \n",
       "2          GLKO  \n",
       "3          GLKO  \n",
       "4          GLKO  \n",
       "5          GLKO  \n",
       "...         ...  \n",
       "1496       PCKO  \n",
       "1497       PCKO  \n",
       "1498       PCKO  \n",
       "1499       PCKO  \n",
       "1500       PCKO  \n",
       "\n",
       "[1500 rows x 8 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dt_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37bc130e-b865-41ae-9a8b-279ddf3e988a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9806c002-9931-4a30-b43d-f70d69af8ad9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0d29935-a229-4fc1-9e2f-d2ccb1c28dae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75b5f9f5-dc28-4db6-a617-8efbbe8aa353",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83dba446-a090-4c03-835b-e4423c3a5f4e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a390ed1-4c7c-47c1-83df-a06faf72517d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bbbe1e27-cd72-4098-ba48-4b8c7423550e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-29 17:03:30,492 [INFO]: R: Astrocytes: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:05:33,488 [INFO]: R: BAMs: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:06:40,545 [INFO]: R: Choroid_Plexus: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:06:45,008 [INFO]: R: ECs: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:06:51,992 [INFO]: R: Ependymal: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:07:26,106 [INFO]: R: Immune_Other: skip (ref_group/test_groups missing)\n",
      "2025-10-29 17:07:26,287 [INFO]: R: Microglia: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:07:57,213 [INFO]: R: Neurons_Dopa: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:08:09,777 [INFO]: R: Neurons_Gaba: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:08:14,478 [INFO]: R: Neurons_Glut: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:08:15,849 [INFO]: R: Neurons_Granule_Immature: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:08:34,967 [INFO]: R: Neurons_Other: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:08:57,551 [INFO]: R: OPCs: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:09:21,694 [INFO]: R: Oligodendrocytes: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:09:25,111 [INFO]: R: Pericytes: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:09:27,764 [INFO]: R: SMCs: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:10:12,755 [INFO]: R: Undefined: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:10:23,625 [INFO]: R: VLMCs: WT_6/WT_12/WT_18/WT_24 vs WT_3\n",
      "2025-10-29 17:10:33,624 [INFO]: R: all: WT_6/WT_12/WT_18/WT_24 vs WT_3\n"
     ]
    }
   ],
   "source": [
    "all_degs = {}\n",
    "for group_i, adata_tmp in adatas_pb.items():\n",
    "    with localconverter(conv):\n",
    "        res = ro.r['dream_loop'](\n",
    "            adata=adata_tmp,\n",
    "            group_i=group_i,\n",
    "            test_groups=ro.StrVector(args.test_groups),\n",
    "            ref_group=args.ref,\n",
    "            condition_col=args.condition_key,\n",
    "            batch_col=(args.batch_key if args.batch_key is not None else ro.NULL),\n",
    "            threads=args.threads,\n",
    "            min_count=2\n",
    "        )\n",
    "    if res is not ro.NULL:\n",
    "        all_degs[group_i] = res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "eb8523a9-a684-40de-828f-47aecc0db2c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-29 17:10:38,823 [INFO]: Format output table...\n",
      "2025-10-29 17:10:39,007 [INFO]: Add ensembl gene ids...\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "3 input query terms found no hit:\t['Pifo', 'Ctps', 'H2afj']\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "2025-10-29 17:10:43,431 [INFO]: 500 unique symbols -> 500 mapped (100.0%)\n",
      "2025-10-29 17:10:43,432 [INFO]: Renamed aliases: Pifo->Cimap3, Ctps->Ctps1, H2afj->H2aj\n"
     ]
    }
   ],
   "source": [
    "logger.info(\"Format output table...\")\n",
    "collapsed_df = (\n",
    "    pd.concat([pd.DataFrame(v).assign(subset=k) for k, v in all_degs.items()],\n",
    "              ignore_index=True)\n",
    "    .set_index(\"gene\", drop=False)\n",
    ")\n",
    "\n",
    "front = [\"subset\", \"gene\", \"FC\", \"logFC\", \"PValue\", \"FDR\", \"AveExpr\", \"t\", \"B\"]\n",
    "order = [c for c in front if c in collapsed_df.columns]\n",
    "rest  = [c for c in collapsed_df.columns if c not in order]\n",
    "collapsed_df = collapsed_df[order + rest]\n",
    "\n",
    "collapsed_df = add_group_sample_counts(\n",
    "    collapsed_df, adatas_pb,\n",
    "    condition_key=args.condition_key,\n",
    "    sample_key=args.sample_key,\n",
    "    ref=args.ref,\n",
    "    test_groups=args.test_groups,\n",
    "    subset_group=\"subset\"\n",
    ")\n",
    "\n",
    "logger.info(\"Add ensembl gene ids...\")\n",
    "collapsed_df = add_ensembl_id(collapsed_df, logger=logger)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5f504f09-47c9-4ef8-a490-1fac1931c56c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subset</th>\n",
       "      <th>gene</th>\n",
       "      <th>FC</th>\n",
       "      <th>logFC</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>t</th>\n",
       "      <th>B</th>\n",
       "      <th>method</th>\n",
       "      <th>test_group</th>\n",
       "      <th>ref</th>\n",
       "      <th>test</th>\n",
       "      <th>test_group_n</th>\n",
       "      <th>ref_n</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>0.312137</td>\n",
       "      <td>-1.679751</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.038272</td>\n",
       "      <td>7.673805</td>\n",
       "      <td>-6.669579</td>\n",
       "      <td>1.572729</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_6</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_6vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000027270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>Gata3</td>\n",
       "      <td>6.032899</td>\n",
       "      <td>2.592851</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>0.038272</td>\n",
       "      <td>3.380014</td>\n",
       "      <td>6.545563</td>\n",
       "      <td>1.366111</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_6</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_6vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000015619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>Mmp9</td>\n",
       "      <td>3.609200</td>\n",
       "      <td>1.851679</td>\n",
       "      <td>0.000855</td>\n",
       "      <td>0.142474</td>\n",
       "      <td>3.138626</td>\n",
       "      <td>5.079559</td>\n",
       "      <td>-0.181532</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_6</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_6vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000017737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>Abcb1a</td>\n",
       "      <td>2.236494</td>\n",
       "      <td>1.161239</td>\n",
       "      <td>0.001944</td>\n",
       "      <td>0.242947</td>\n",
       "      <td>7.128629</td>\n",
       "      <td>4.455991</td>\n",
       "      <td>-1.055956</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_6</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_6vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000040584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Astrocytes</td>\n",
       "      <td>Bcl2</td>\n",
       "      <td>1.922956</td>\n",
       "      <td>0.943326</td>\n",
       "      <td>0.005291</td>\n",
       "      <td>0.506050</td>\n",
       "      <td>7.619215</td>\n",
       "      <td>3.747909</td>\n",
       "      <td>-2.083698</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_6</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_6vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000057329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35995</th>\n",
       "      <td>all</td>\n",
       "      <td>Tgfb3</td>\n",
       "      <td>0.997319</td>\n",
       "      <td>-0.003873</td>\n",
       "      <td>0.977760</td>\n",
       "      <td>0.985645</td>\n",
       "      <td>10.225112</td>\n",
       "      <td>-0.028501</td>\n",
       "      <td>-6.738243</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_24</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_24vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000021253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35996</th>\n",
       "      <td>all</td>\n",
       "      <td>Gatm</td>\n",
       "      <td>1.001780</td>\n",
       "      <td>0.002566</td>\n",
       "      <td>0.987958</td>\n",
       "      <td>0.993390</td>\n",
       "      <td>14.624825</td>\n",
       "      <td>0.015431</td>\n",
       "      <td>-6.841523</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_24</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_24vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000027199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35997</th>\n",
       "      <td>all</td>\n",
       "      <td>C1qtnf4</td>\n",
       "      <td>0.998493</td>\n",
       "      <td>-0.002176</td>\n",
       "      <td>0.989417</td>\n",
       "      <td>0.993390</td>\n",
       "      <td>12.965992</td>\n",
       "      <td>-0.013561</td>\n",
       "      <td>-6.834521</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_24</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_24vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000040794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35998</th>\n",
       "      <td>all</td>\n",
       "      <td>Prss23</td>\n",
       "      <td>0.999370</td>\n",
       "      <td>-0.000910</td>\n",
       "      <td>0.995129</td>\n",
       "      <td>0.995203</td>\n",
       "      <td>10.660097</td>\n",
       "      <td>-0.006241</td>\n",
       "      <td>-6.752030</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_24</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_24vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000039405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35999</th>\n",
       "      <td>all</td>\n",
       "      <td>Jam2</td>\n",
       "      <td>0.999396</td>\n",
       "      <td>-0.000871</td>\n",
       "      <td>0.995203</td>\n",
       "      <td>0.995203</td>\n",
       "      <td>11.908866</td>\n",
       "      <td>-0.006146</td>\n",
       "      <td>-6.823172</td>\n",
       "      <td>DREAM</td>\n",
       "      <td>WT_24</td>\n",
       "      <td>WT_3</td>\n",
       "      <td>WT_24vsWT_3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000053062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36000 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           subset     gene        FC     logFC    PValue       FDR    AveExpr  \\\n",
       "0      Astrocytes    Lamp5  0.312137 -1.679751  0.000134  0.038272   7.673805   \n",
       "1      Astrocytes    Gata3  6.032899  2.592851  0.000153  0.038272   3.380014   \n",
       "2      Astrocytes     Mmp9  3.609200  1.851679  0.000855  0.142474   3.138626   \n",
       "3      Astrocytes   Abcb1a  2.236494  1.161239  0.001944  0.242947   7.128629   \n",
       "4      Astrocytes     Bcl2  1.922956  0.943326  0.005291  0.506050   7.619215   \n",
       "...           ...      ...       ...       ...       ...       ...        ...   \n",
       "35995         all    Tgfb3  0.997319 -0.003873  0.977760  0.985645  10.225112   \n",
       "35996         all     Gatm  1.001780  0.002566  0.987958  0.993390  14.624825   \n",
       "35997         all  C1qtnf4  0.998493 -0.002176  0.989417  0.993390  12.965992   \n",
       "35998         all   Prss23  0.999370 -0.000910  0.995129  0.995203  10.660097   \n",
       "35999         all     Jam2  0.999396 -0.000871  0.995203  0.995203  11.908866   \n",
       "\n",
       "              t         B method test_group   ref         test  test_group_n  \\\n",
       "0     -6.669579  1.572729  DREAM       WT_6  WT_3   WT_6vsWT_3             3   \n",
       "1      6.545563  1.366111  DREAM       WT_6  WT_3   WT_6vsWT_3             3   \n",
       "2      5.079559 -0.181532  DREAM       WT_6  WT_3   WT_6vsWT_3             3   \n",
       "3      4.455991 -1.055956  DREAM       WT_6  WT_3   WT_6vsWT_3             3   \n",
       "4      3.747909 -2.083698  DREAM       WT_6  WT_3   WT_6vsWT_3             3   \n",
       "...         ...       ...    ...        ...   ...          ...           ...   \n",
       "35995 -0.028501 -6.738243  DREAM      WT_24  WT_3  WT_24vsWT_3             3   \n",
       "35996  0.015431 -6.841523  DREAM      WT_24  WT_3  WT_24vsWT_3             3   \n",
       "35997 -0.013561 -6.834521  DREAM      WT_24  WT_3  WT_24vsWT_3             3   \n",
       "35998 -0.006241 -6.752030  DREAM      WT_24  WT_3  WT_24vsWT_3             3   \n",
       "35999 -0.006146 -6.823172  DREAM      WT_24  WT_3  WT_24vsWT_3             3   \n",
       "\n",
       "       ref_n          ensembl_id  \n",
       "0          3  ENSMUSG00000027270  \n",
       "1          3  ENSMUSG00000015619  \n",
       "2          3  ENSMUSG00000017737  \n",
       "3          3  ENSMUSG00000040584  \n",
       "4          3  ENSMUSG00000057329  \n",
       "...      ...                 ...  \n",
       "35995      3  ENSMUSG00000021253  \n",
       "35996      3  ENSMUSG00000027199  \n",
       "35997      3  ENSMUSG00000040794  \n",
       "35998      3  ENSMUSG00000039405  \n",
       "35999      3  ENSMUSG00000053062  \n",
       "\n",
       "[36000 rows x 16 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "collapsed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "051b6f2f-71f2-4259-9815-68f22032fffe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args.overwrite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5de610ee-02ab-49ef-9a4f-b46f0a7281e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-29 17:10:43,599 [INFO]: Export table(s)...\n",
      "2025-10-29 17:10:46,119 [INFO]:   Wrote: /dss/dssfs03/pn52re/pn52re-dss-0001/cellseg-benchmark/analysis/aging/Cellpose_1_nuclei_model/dea/aging-by-celltype_pseudobulk_DREAM_WT_12vsWT_3.xlsx\n",
      "2025-10-29 17:10:48,699 [INFO]:   Wrote: /dss/dssfs03/pn52re/pn52re-dss-0001/cellseg-benchmark/analysis/aging/Cellpose_1_nuclei_model/dea/aging-by-celltype_pseudobulk_DREAM_WT_18vsWT_3.xlsx\n",
      "2025-10-29 17:10:50,938 [INFO]:   Wrote: /dss/dssfs03/pn52re/pn52re-dss-0001/cellseg-benchmark/analysis/aging/Cellpose_1_nuclei_model/dea/aging-by-celltype_pseudobulk_DREAM_WT_24vsWT_3.xlsx\n",
      "2025-10-29 17:10:53,844 [INFO]:   Wrote: /dss/dssfs03/pn52re/pn52re-dss-0001/cellseg-benchmark/analysis/aging/Cellpose_1_nuclei_model/dea/aging-by-celltype_pseudobulk_DREAM_WT_6vsWT_3.xlsx\n",
      "2025-10-29 17:10:53,853 [INFO]: Done.\n"
     ]
    }
   ],
   "source": [
    "logger.info(\"Export table(s)...\")\n",
    "subset_key_clean = re.sub(r'[_-]', '', args.subset_key)\n",
    "name = f\"{args.cohort}-by-{subset_key_clean}_pseudobulk\"\n",
    "\n",
    "for (method, test), df_mt in collapsed_df.groupby([\"method\", \"test\"]):\n",
    "    df_mt = df_mt.loc[:, df_mt.notna().any()]  # drop all-NaN cols\n",
    "    xlsx = output_dir / f\"{name}_{method}_{test}.xlsx\"\n",
    "    if xlsx.exists() and not args.overwrite:\n",
    "        logger.info(f\"  Exists, skip: {xlsx}\")\n",
    "        continue\n",
    "    used = set()\n",
    "    with pd.ExcelWriter(xlsx, engine=\"xlsxwriter\") as writer:\n",
    "        for gid, g in df_mt.groupby(\"subset\", sort=True):\n",
    "            g.sort_values(\"PValue\").to_excel(\n",
    "                writer, sheet_name=safe_sheet(gid, used), index=False\n",
    "            )\n",
    "    logger.info(f\"  Wrote: {xlsx}\")\n",
    "        \n",
    "logger.info(f\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f98c1a65-7f3f-4c38-b074-4a67a4948210",
   "metadata": {},
   "outputs": [],
   "source": [
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0dcd018-811c-4676-a1ed-41ee75b18fdd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
