{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9144e11a-ad08-4b49-9420-3d9e3b5d5b9c",
   "metadata": {},
   "source": [
    "# Morphology properties\n",
    "* total transcripts\n",
    "* cell area / volume\n",
    "* cell roundness\n",
    "* cell elongation\n",
    "* DAPI/PolyT brightness\n",
    "* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "20b5b882-bb4a-48f9-8163-44c911aeccd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cellseg_benchmark.metrics import read_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6824a98e-5f25-496e-8c7f-93f5979dc39a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading adata for method Proseg_3D_Cellpose_1_nuclei_model\n"
     ]
    }
   ],
   "source": [
    "adata = read_adata(\"aging\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6ed5486a-7112-49f7-8c75-d37470eab1ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['cell', 'original_cell_id', 'centroid_x', 'centroid_y', 'centroid_z',\n",
       "       'component', 'volume', 'surface_area', 'scale', 'region', 'slide',\n",
       "       'cell_type_incl_low_quality_revised',\n",
       "       'cell_type_mmc_incl_low_quality_clusters',\n",
       "       'cell_type_mmc_incl_low_quality', 'cell_type_incl_mixed_revised',\n",
       "       'cell_type_mmc_incl_mixed_clusters', 'cell_type_mmc_incl_mixed',\n",
       "       'cell_type_revised', 'cell_type_mmc_raw_clusters', 'cell_type_mmc_raw',\n",
       "       'area', 'volume_sum', 'volume_trapz', 'volume_final', 'num_z_planes',\n",
       "       'size_normalized', 'surface_to_volume_ratio', 'sphericity', 'solidity',\n",
       "       'elongation', 'genotype', 'age_months', 'run_date', 'organism',\n",
       "       'cohort', 'sample', 'animal_id', 'condition', 'n_counts', 'n_genes',\n",
       "       'path', 'spatial_outlier', 'low_quality_cell', 'volume_outlier_cell',\n",
       "       'fov'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5df1cc38-c482-4d3b-847e-3b6afc99177b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_integrity</th>\n",
       "      <th>variance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aaaaaaaa-1</th>\n",
       "      <td>0.821117</td>\n",
       "      <td>0.456002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaaaaaab-1</th>\n",
       "      <td>0.790138</td>\n",
       "      <td>0.331503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaaaaaac-1</th>\n",
       "      <td>0.801383</td>\n",
       "      <td>0.708477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaaaaaad-1</th>\n",
       "      <td>0.874341</td>\n",
       "      <td>0.437607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaaaaaae-1</th>\n",
       "      <td>0.820528</td>\n",
       "      <td>0.473899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaaalfma-1-9</th>\n",
       "      <td>0.639256</td>\n",
       "      <td>0.317012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaaalfmb-1-9</th>\n",
       "      <td>0.608563</td>\n",
       "      <td>0.217885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaaalfme-1-9</th>\n",
       "      <td>0.530731</td>\n",
       "      <td>0.280629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaaalfmf-1-9</th>\n",
       "      <td>0.574817</td>\n",
       "      <td>0.211674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaaalfmg-1-9</th>\n",
       "      <td>0.568003</td>\n",
       "      <td>0.288153</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>667954 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              mean_integrity  variance\n",
       "aaaaaaaa-1          0.821117  0.456002\n",
       "aaaaaaab-1          0.790138  0.331503\n",
       "aaaaaaac-1          0.801383  0.708477\n",
       "aaaaaaad-1          0.874341  0.437607\n",
       "aaaaaaae-1          0.820528  0.473899\n",
       "...                      ...       ...\n",
       "aaaalfma-1-9        0.639256  0.317012\n",
       "aaaalfmb-1-9        0.608563  0.217885\n",
       "aaaalfme-1-9        0.530731  0.280629\n",
       "aaaalfmf-1-9        0.574817  0.211674\n",
       "aaaalfmg-1-9        0.568003  0.288153\n",
       "\n",
       "[667954 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obsm[\"Ovrlpy_stats\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1db93f22-6e9b-4096-8c31-5519cc71b167",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/dask/dataframe/__init__.py:31: FutureWarning: The legacy Dask DataFrame implementation is deprecated and will be removed in a future version. Set the configuration option `dataframe.query-planning` to `True` or None to enable the new Dask Dataframe implementation and silence this warning.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import spatialdata as sd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "06aa960b-66f6-44b4-8763-b0faeff2a331",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lrwxrwxrwx 1 ra64luf2 pn52go 53 Aug 27 19:11 /home/ubuntu/gitrepos/cellseg-benchmark/data -> /dss/dssfs03/pn52re/pn52re-dss-0001/cellseg-benchmark\n"
     ]
    }
   ],
   "source": [
    "data_dir = os.path.abspath(\"/home/ubuntu/gitrepos/cellseg-benchmark/data\")\n",
    "!ls -l $data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b6ab78cc-37ea-4d1e-a9bf-452c511d879d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/zarr/creation.py:610: UserWarning: ignoring keyword argument 'read_only'\n",
      "  compressor, fill_value = _kwargs_compat(compressor, fill_value, kwargs)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/zarr/creation.py:610: UserWarning: ignoring keyword argument 'read_only'\n",
      "  compressor, fill_value = _kwargs_compat(compressor, fill_value, kwargs)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/zarr/creation.py:610: UserWarning: ignoring keyword argument 'read_only'\n",
      "  compressor, fill_value = _kwargs_compat(compressor, fill_value, kwargs)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/zarr/creation.py:610: UserWarning: ignoring keyword argument 'read_only'\n",
      "  compressor, fill_value = _kwargs_compat(compressor, fill_value, kwargs)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/zarr/creation.py:610: UserWarning: ignoring keyword argument 'read_only'\n",
      "  compressor, fill_value = _kwargs_compat(compressor, fill_value, kwargs)\n"
     ]
    }
   ],
   "source": [
    "sample_name = \"foxf2_s2_r1\"\n",
    "sdata_path = os.path.join(data_dir, \"samples\", sample_name, \"sdata_z3.zarr\")\n",
    "if \"sdata\" not in locals():\n",
    "    sdata = sd.read_zarr(sdata_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e501975-3052-4061-889e-9c9c1b78641e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if sdata points table contains any annotation of boundaries. Check if can compute this.\n",
    "# need to get stats of number of transcripts overall, and number of transcripts per gene\n",
    "# check if this info is also in adata.var already (for comparing annotated transcripts per gene)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
