{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "42706144-9bb6-4cdc-8200-ca4676279e31",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/dask/dataframe/__init__.py:31: FutureWarning: The legacy Dask DataFrame implementation is deprecated and will be removed in a future version. Set the configuration option `dataframe.query-planning` to `True` or None to enable the new Dask Dataframe implementation and silence this warning.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "import logging\n",
    "import os\n",
    "import sys\n",
    "import warnings\n",
    "from concurrent.futures import ProcessPoolExecutor, as_completed\n",
    "from pathlib import Path\n",
    "from typing import Tuple\n",
    "\n",
    "import scanpy as sc\n",
    "import yaml\n",
    "from anndata import AnnData\n",
    "from spatialdata import read_zarr\n",
    "\n",
    "sys.path.insert(1, \"/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark\")\n",
    "from cellseg_benchmark.adata_utils import (\n",
    "    filter_genes,\n",
    "    filter_low_quality_cells,\n",
    "    filter_spatial_outlier_cells,\n",
    "    integration_harmony,\n",
    "    merge_adatas,\n",
    "    normalize_counts,\n",
    "    pca_umap_single,\n",
    ")\n",
    "\n",
    "\n",
    "def _load_one(\n",
    "    sample_dir: Path, seg_method: str, logger: logging.Logger\n",
    ") -> Tuple[str, AnnData | None]:\n",
    "    \"\"\"Load AnnData from one master sdata.\"\"\"\n",
    "    sdata = read_zarr(sample_dir / \"sdata_z3.zarr\", selection=(\"tables\",))\n",
    "    if f\"adata_{seg_method}\" not in sdata.tables.keys():\n",
    "        if logger:\n",
    "            logger.warning(f\"Skipping {seg_method}. No such key: {seg_method}\")\n",
    "        return sample_dir.name, None\n",
    "    return sample_dir.name, sdata[f\"adata_{seg_method}\"]\n",
    "\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", \".*The table is annotating*\", UserWarning)\n",
    "\n",
    "sc.settings.n_jobs = -1\n",
    "\n",
    "# Logger setup\n",
    "logger = logging.getLogger(\"integrate_adatas\")\n",
    "logger.setLevel(logging.INFO)\n",
    "handler = logging.StreamHandler()\n",
    "handler.setFormatter(logging.Formatter(\"%(asctime)s [%(levelname)s]: %(message)s\"))\n",
    "logger.addHandler(handler)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1542ea05-f161-4b7b-878a-e3c254115420",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Namespace(cohort='foxf2', seg_method='Cellpose_1_nuclei_model')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CLI args\n",
    "sys.argv = [\"notebook\", \"foxf2\", \"Cellpose_1_nuclei_model\"]\n",
    "# sys.argv = [\"notebook\", \"foxf2\", \"Proseg_Cellpose_1_nuclei_model\"]\n",
    "\n",
    "parser = argparse.ArgumentParser(\n",
    "    description=\"Integrate adatas from a selected segmentation method.\"\n",
    ")\n",
    "parser.add_argument(\"cohort\", help=\"Cohort name, e.g., 'foxf2'\")\n",
    "parser.add_argument(\n",
    "    \"seg_method\", help=\"Segmentation method, e.g., 'Cellpose_1_nuclei_model'\"\n",
    ")\n",
    "args = parser.parse_args()\n",
    "args"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ae8601a0-daa8-49f9-afaf-be9b2adb9abd",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = Path(\"/dss/dssfs03/pn52re/pn52re-dss-0001/cellseg-benchmark\")\n",
    "samples_path = base_path / \"samples\"\n",
    "save_path = base_path / \"analysis\" / args.cohort / args.seg_method\n",
    "save_path.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b727c227-5b82-4be9-a66d-54db0ed5757e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'foxf2_s1_r0': {'cohort': 'foxf2',\n",
       "  'slide': 1,\n",
       "  'region': 0,\n",
       "  'genotype': 'ECKO',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240229',\n",
       "  'animal_id': '000',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240229_Foxf2-Slide01-cp-WT-ECKO/region_0-ECKO000'},\n",
       " 'foxf2_s1_r1': {'cohort': 'foxf2',\n",
       "  'slide': 1,\n",
       "  'region': 1,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240229',\n",
       "  'animal_id': '000',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240229_Foxf2-Slide01-cp-WT-ECKO/region_1-WT000'},\n",
       " 'foxf2_s2_r1': {'cohort': 'foxf2',\n",
       "  'slide': 2,\n",
       "  'region': 1,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240322',\n",
       "  'animal_id': '536',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240322_Foxf2-Slide02-cp-WT-PCKO/region_1-WT536'},\n",
       " 'foxf2_s2_r2': {'cohort': 'foxf2',\n",
       "  'slide': 2,\n",
       "  'region': 2,\n",
       "  'genotype': 'PCKO',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240322',\n",
       "  'animal_id': '239',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240322_Foxf2-Slide02-cp-WT-PCKO/region_2-PCKO239'},\n",
       " 'foxf2_s4_r0': {'cohort': 'foxf2',\n",
       "  'slide': 4,\n",
       "  'region': 0,\n",
       "  'genotype': 'PCKO',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240304',\n",
       "  'animal_id': '421',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240304_Foxf2-Slide04-cp-ECKO-PCKO/region_0-PCKO421'},\n",
       " 'foxf2_s4_r1': {'cohort': 'foxf2',\n",
       "  'slide': 4,\n",
       "  'region': 1,\n",
       "  'genotype': 'ECKO',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240304',\n",
       "  'animal_id': '557',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240304_Foxf2-Slide04-cp-ECKO-PCKO/region_1-ECKO557'},\n",
       " 'foxf2_s5_r0': {'cohort': 'foxf2',\n",
       "  'slide': 5,\n",
       "  'region': 0,\n",
       "  'genotype': 'PCKO',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240628',\n",
       "  'animal_id': '238',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240628_Foxf2-Slide05-cp-PCKO-ECKO/region_0-PCKO238'},\n",
       " 'foxf2_s5_r1': {'cohort': 'foxf2',\n",
       "  'slide': 5,\n",
       "  'region': 1,\n",
       "  'genotype': 'ECKO',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240628',\n",
       "  'animal_id': '544',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240628_Foxf2-Slide05-cp-PCKO-ECKO/region_1-ECKO544'},\n",
       " 'foxf2_s6_r0': {'cohort': 'foxf2',\n",
       "  'slide': 6,\n",
       "  'region': 0,\n",
       "  'genotype': 'GLKO',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240729',\n",
       "  'animal_id': '890',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240729_Foxf2-Slide06-cp-WT-GLKO-GLKO/region_0-KO890'},\n",
       " 'foxf2_s6_r1': {'cohort': 'foxf2',\n",
       "  'slide': 6,\n",
       "  'region': 1,\n",
       "  'genotype': 'GLKO',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240729',\n",
       "  'animal_id': '883',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240729_Foxf2-Slide06-cp-WT-GLKO-GLKO/region_1-KO883'},\n",
       " 'foxf2_s6_r2': {'cohort': 'foxf2',\n",
       "  'slide': 6,\n",
       "  'region': 2,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240729',\n",
       "  'animal_id': '897',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240729_Foxf2-Slide06-cp-WT-GLKO-GLKO/region_2-WT897'},\n",
       " 'foxf2_s7_r0': {'cohort': 'foxf2',\n",
       "  'slide': 7,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240805',\n",
       "  'animal_id': '886',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240805_Foxf2-Slide07-cp-WT-GLKO/region_0-WT886'},\n",
       " 'foxf2_s7_r1': {'cohort': 'foxf2',\n",
       "  'slide': 7,\n",
       "  'region': 1,\n",
       "  'genotype': 'GLKO',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240805',\n",
       "  'animal_id': '885',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202402-Foxf2/merfish_output/20240805_Foxf2-Slide07-cp-WT-GLKO/region_1-KO885'},\n",
       " 'aging_s1_r0': {'cohort': 'aging',\n",
       "  'slide': 1,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20240516',\n",
       "  'animal_id': '308',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20240516_Htra1-Slide01-cp-WT308-KO307/region_0-WT308_6m'},\n",
       " 'aging_s5_r0': {'cohort': 'aging',\n",
       "  'slide': 5,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 18,\n",
       "  'run_date': '20241220',\n",
       "  'animal_id': '1377',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20241220_Htra1-and-Aging-Slide05-cp-WT1377_18m-WT1332_6m-WT_eNOS_12m/region_0-WT1377_18m'},\n",
       " 'aging_s5_r1': {'cohort': 'aging',\n",
       "  'slide': 5,\n",
       "  'region': 1,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20241220',\n",
       "  'animal_id': '1332',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20241220_Htra1-and-Aging-Slide05-cp-WT1377_18m-WT1332_6m-WT_eNOS_12m/region_1-WT1332_6m'},\n",
       " 'aging_s5_r2': {'cohort': 'aging',\n",
       "  'slide': 5,\n",
       "  'region': 2,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 12,\n",
       "  'run_date': '20241220',\n",
       "  'animal_id': '2',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20241220_Htra1-and-Aging-Slide05-cp-WT1377_18m-WT1332_6m-WT_eNOS_12m/region_2-WT2eNos_12m'},\n",
       " 'aging_s6_r0': {'cohort': 'aging',\n",
       "  'slide': 6,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 18,\n",
       "  'run_date': '20250113',\n",
       "  'animal_id': '421',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250113_Htra1-and-Aging-Slide06-cp-WT421_18m-KO278_18m-KO276_18m/region_0-WT421_18m'},\n",
       " 'aging_s7_r2': {'cohort': 'aging',\n",
       "  'slide': 7,\n",
       "  'region': 2,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 6,\n",
       "  'run_date': '20250127',\n",
       "  'animal_id': '58',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250127_Htra1-Slide07-cp-TG79-WT58-rerunSlide02/region_2-WT58_6m'},\n",
       " 'aging_s8_r0': {'cohort': 'aging',\n",
       "  'slide': 8,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 12,\n",
       "  'run_date': '20250203',\n",
       "  'animal_id': '1',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250203_Aging-Slide08-cp-WT_eNos_12m-WT1377_18m-WT121_3m/region_0-WT1eNos_12m'},\n",
       " 'aging_s8_r1': {'cohort': 'aging',\n",
       "  'slide': 8,\n",
       "  'region': 1,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 3,\n",
       "  'run_date': '20250203',\n",
       "  'animal_id': '121',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250203_Aging-Slide08-cp-WT_eNos_12m-WT1377_18m-WT121_3m/region_1-WT121_3m'},\n",
       " 'aging_s8_r2': {'cohort': 'aging',\n",
       "  'slide': 8,\n",
       "  'region': 2,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 18,\n",
       "  'run_date': '20250203',\n",
       "  'animal_id': '1337',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250203_Aging-Slide08-cp-WT_eNos_12m-WT1377_18m-WT121_3m/region_2-WT1337_18m'},\n",
       " 'aging_s10_r0': {'cohort': 'aging',\n",
       "  'slide': 10,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 24,\n",
       "  'run_date': '20250508',\n",
       "  'animal_id': '132',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250508_Aging-Slide10-cp-WT308_3m-WT309_3m-WT132_24m/region_0-WT132_24m'},\n",
       " 'aging_s10_r1': {'cohort': 'aging',\n",
       "  'slide': 10,\n",
       "  'region': 1,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 3,\n",
       "  'run_date': '20250508',\n",
       "  'animal_id': '308',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250508_Aging-Slide10-cp-WT308_3m-WT309_3m-WT132_24m/region_1-WT308_3m'},\n",
       " 'aging_s10_r2': {'cohort': 'aging',\n",
       "  'slide': 10,\n",
       "  'region': 2,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 3,\n",
       "  'run_date': '20250508',\n",
       "  'animal_id': '309',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250508_Aging-Slide10-cp-WT308_3m-WT309_3m-WT132_24m/region_2-WT309_3m'},\n",
       " 'aging_s11_r0': {'cohort': 'aging',\n",
       "  'slide': 11,\n",
       "  'region': 3,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 18,\n",
       "  'run_date': '20250526',\n",
       "  'animal_id': '232',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250526_Aging-Slide11-cp-WT232_18m-WT999_24m-WT888_24m/region_3-WT232_18m'},\n",
       " 'aging_s11_r1': {'cohort': 'aging',\n",
       "  'slide': 11,\n",
       "  'region': 1,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 24,\n",
       "  'run_date': '20250526',\n",
       "  'animal_id': '999',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250526_Aging-Slide11-cp-WT232_18m-WT999_24m-WT888_24m/region_1-WT999_24m'},\n",
       " 'aging_s11_r2': {'cohort': 'aging',\n",
       "  'slide': 11,\n",
       "  'region': 2,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 24,\n",
       "  'run_date': '20250526',\n",
       "  'animal_id': '888',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250526_Aging-Slide11-cp-WT232_18m-WT999_24m-WT888_24m/region_2-WT888_24m'},\n",
       " 'aging_s12_r0': {'cohort': 'aging',\n",
       "  'slide': 12,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': 12,\n",
       "  'run_date': '20250623',\n",
       "  'animal_id': '279',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0000/202405-Htra1-and-Aging/merfish_output/20250623_Aging-Slide12-cp-WT279_12m/region_0-WT279_12m'},\n",
       " 'VizgenMouseBrain_s1_r0': {'cohort': 'VizgenMouseBrain',\n",
       "  'slide': 1,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20210501',\n",
       "  'animal_id': 'Vizgen1',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/Vizgen-BrainReceptorShowcase-500plex/Slice2/Replicate1'},\n",
       " 'VizgenMouseBrain_s2_r0': {'cohort': 'VizgenMouseBrain',\n",
       "  'slide': 2,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20210501',\n",
       "  'animal_id': 'Vizgen1',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/Vizgen-BrainReceptorShowcase-500plex/Slice3/Replicate1'},\n",
       " 'VizgenMouseBrain_s3_r0': {'cohort': 'VizgenMouseBrain',\n",
       "  'slide': 3,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20210501',\n",
       "  'animal_id': 'Vizgen2',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/Vizgen-BrainReceptorShowcase-500plex/Slice2/Replicate2'},\n",
       " 'VizgenMouseBrain_s4_r0': {'cohort': 'VizgenMouseBrain',\n",
       "  'slide': 4,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20210501',\n",
       "  'animal_id': 'Vizgen2',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/Vizgen-BrainReceptorShowcase-500plex/Slice3/Replicate2'},\n",
       " 'VizgenMouseBrain_s5_r0': {'cohort': 'VizgenMouseBrain',\n",
       "  'slide': 5,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20210501',\n",
       "  'animal_id': 'Vizgen3',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/Vizgen-BrainReceptorShowcase-500plex/Slice2/Replicate3'},\n",
       " 'VizgenMouseBrain_s6_r0': {'cohort': 'VizgenMouseBrain',\n",
       "  'slide': 6,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20210501',\n",
       "  'animal_id': 'Vizgen3',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/Vizgen-BrainReceptorShowcase-500plex/Slice3/Replicate3'},\n",
       " 'ABCAtlas_s1_r0': {'cohort': 'ABCAtlas',\n",
       "  'slide': 1,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20220906',\n",
       "  'animal_id': '638850',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/ABCAtlas-MouseBrain-500plex/1199650959'},\n",
       " 'ABCAtlas_s2_r0': {'cohort': 'ABCAtlas',\n",
       "  'slide': 2,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20220929',\n",
       "  'animal_id': '638850',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/ABCAtlas-MouseBrain-500plex/1199651002'},\n",
       " 'ABCAtlas_s3_r0': {'cohort': 'ABCAtlas',\n",
       "  'slide': 3,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20220916',\n",
       "  'animal_id': '638850',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/ABCAtlas-MouseBrain-500plex/1199651036'},\n",
       " 'ABCAtlas_s4_r0': {'cohort': 'ABCAtlas',\n",
       "  'slide': 4,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20220919',\n",
       "  'animal_id': '638850',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/ABCAtlas-MouseBrain-500plex/1199651048'},\n",
       " 'ABCAtlas_s5_r0': {'cohort': 'ABCAtlas',\n",
       "  'slide': 5,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20220922',\n",
       "  'animal_id': '638850',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/ABCAtlas-MouseBrain-500plex/1199651078'},\n",
       " 'ABCAtlas_s6_r0': {'cohort': 'ABCAtlas',\n",
       "  'slide': 6,\n",
       "  'region': 0,\n",
       "  'genotype': 'WT',\n",
       "  'age_months': None,\n",
       "  'run_date': '20220927',\n",
       "  'animal_id': '638850',\n",
       "  'organism': 'mouse',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/ABCAtlas-MouseBrain-500plex/1199651112'},\n",
       " 'VizgenHumanBrain_s1_r0': {'cohort': 'VizgenHumanBrain',\n",
       "  'slide': 1,\n",
       "  'region': 0,\n",
       "  'genotype': None,\n",
       "  'age_months': None,\n",
       "  'run_date': '20240801',\n",
       "  'animal_id': None,\n",
       "  'organism': 'human',\n",
       "  'path': '/dss/dssfs03/pn52re/pn52re-dss-0001/merfish-public-data/Vizgen-HumanBrain-AD-1000plex/region_0'}}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_metadata_file, excluded = (\n",
    "    yaml.safe_load(open(base_path / \"misc\" / f))\n",
    "    for f in [\"sample_metadata.yaml\", \"samples_excluded.yaml\"]\n",
    ")  ###############\n",
    "sample_metadata_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "264f5c9b-9b31-4de4-bbe8-a3d93f0abc78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'foxf2_s2_r0', 'foxf2_s3_r0', 'foxf2_s3_r1'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "excluded_samples = set(excluded.get(args.cohort, []))  ###############\n",
    "excluded_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "415d5c79-0ca3-44e5-986e-976e4f93bd1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-17 09:48:03,935 [INFO]: Loading data...\n"
     ]
    }
   ],
   "source": [
    "# Load sdata\n",
    "logger.info(\"Loading data...\")\n",
    "loads = []\n",
    "for sample_dir in samples_path.glob(f\"{args.cohort}*\"):  ###############\n",
    "    if sample_dir.name in excluded_samples:\n",
    "        continue\n",
    "    if not (sample_dir / \"sdata_z3.zarr\").exists():\n",
    "        logger.error(\"master sdata in %s not found.\", sample_dir)\n",
    "        continue\n",
    "    loads.append(sample_dir)\n",
    "\n",
    "# drop samples where AnnData could not be loaded (missing or invalid), e.g. aging s8 r1 Cellpose 2 Transcripts\n",
    "with ProcessPoolExecutor(max_workers=int(os.getenv(\"SLURM_CPUS_PER_TASK\", 1))) as ex:\n",
    "    futures = {ex.submit(_load_one, p, args.seg_method, logger): p for p in loads}\n",
    "    adata_list = [f.result() for f in as_completed(futures)]\n",
    "adata_list = [(x, y) for x, y in adata_list if y is not None]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1263653e-206a-41ee-8215-c64f0964e769",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-17 09:53:16,658 [INFO]: Merging adatas of Cellpose_1_nuclei_model\n",
      "100%|██████████| 13/13 [00:00<00:00, 40.39it/s]\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1774: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "2025-09-17 09:53:17,985 [INFO]: Cellpose_1_nuclei_model: # of cells: 595581, # of samples: 13\n",
      "2025-09-17 09:53:18,010 [INFO]: Cellpose_1_nuclei_model: # of cells: 595581, # of samples: 13\n",
      "2025-09-17 09:53:18,017 [INFO]: Plotting QC results\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:197: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout()\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/anndata/_core/anndata.py:1158: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:245: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout()\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:259: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_tmp.var[\"n_counts\"] = adata_tmp.X.sum(axis=0).T\n",
      "/dss/dsshome1/0C/ra98gaq/Git/cellseg-benchmark/cellseg_benchmark/adata_utils.py:286: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout()\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ubuntu/miniforge3/lib/python3.12/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    }
   ],
   "source": [
    "# Merge and process\n",
    "adata = merge_adatas(\n",
    "    adata_list,\n",
    "    seg_method=args.seg_method,\n",
    "    logger=logger,\n",
    "    plot_qc_stats=True,\n",
    "    save_path=save_path / \"plots\",\n",
    ")\n",
    "del adata_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d2599ce8-d68d-4d03-8792-bf54ac74200e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# workaround to fix adata.obs formatting ###############\n",
    "adata.obs[\"sample\"] = adata.obs[\"sample\"].str.replace(\n",
    "    rf\"^{args.cohort}_(\\d+)_(\\d+)$\", rf\"{args.cohort}_s\\1_r\\2\", regex=True\n",
    ")\n",
    "adata.obs.drop(columns=[\"spt_region\"], inplace=True, errors=\"ignore\")\n",
    "adata.obs[\"condition\"] = (\n",
    "    adata.obs[\"genotype\"].astype(str) + \"_\" + adata.obs[\"age_months\"].astype(str)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fd42ea73-a56b-460b-8af3-ca5d64204aaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample\n",
       "foxf2_s5_r1    58868\n",
       "foxf2_s5_r0    56940\n",
       "foxf2_s4_r0    49649\n",
       "foxf2_s2_r1    47686\n",
       "foxf2_s6_r2    47030\n",
       "foxf2_s1_r0    46814\n",
       "foxf2_s2_r2    45477\n",
       "foxf2_s6_r1    44109\n",
       "foxf2_s1_r1    42513\n",
       "foxf2_s7_r0    42204\n",
       "foxf2_s7_r1    41277\n",
       "foxf2_s6_r0    37142\n",
       "foxf2_s4_r1    35872\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"sample\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3d16c8de-a18f-40b3-b34b-1d12f602f94e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "condition\n",
       "WT_6      179433\n",
       "PCKO_6    152066\n",
       "ECKO_6    141554\n",
       "GLKO_6    122528\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"condition\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "78e83a0d-83fe-4c01-9fa9-9b27aeea4d3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['region', 'slide', 'cell_type_mmc_incl_low_quality_revised',\n",
       "       'cell_type_mmc_incl_low_quality_clusters',\n",
       "       'cell_type_mmc_incl_low_quality', 'cell_type_mmc_incl_mixed_revised',\n",
       "       'cell_type_mmc_incl_mixed_clusters', 'cell_type_mmc_incl_mixed',\n",
       "       'cell_type_mmc_raw_revised', 'cell_type_mmc_raw_clusters',\n",
       "       'cell_type_mmc_raw', 'area', 'volume_sum', 'volume_final',\n",
       "       'num_z_planes', 'size_normalized', 'surface_to_volume_ratio',\n",
       "       'sphericity', 'solidity', 'elongation', 'genotype', 'age_months',\n",
       "       'run_date', 'animal_id', 'organism', 'cohort', 'sample', 'n_counts',\n",
       "       'n_genes', 'condition'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e601c6b7-6ea9-427d-af0c-03670fc18fad",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-17 09:54:24,996 [INFO]: # total cells before filtering spatial outliers: 595581\n",
      "2025-09-17 09:54:24,997 [INFO]: # spatial_outlier:       13221\n",
      "2025-09-17 09:54:26,290 [INFO]: # total cells after filtering spatial outliers:  582360\n"
     ]
    }
   ],
   "source": [
    "adata.obsm[\"spatial\"] = adata.obsm.get(\"spatial_microns\", adata.obsm[\"spatial\"])\n",
    "adata = filter_spatial_outlier_cells(\n",
    "    adata,\n",
    "    data_dir=str(base_path),\n",
    "    sample_metadata_file=sample_metadata_file,  ############\n",
    "    save_path=save_path / \"plots\",\n",
    "    logger=logger,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a8760c3b-ac78-4e24-a28c-2a4db5a63abf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-17 09:54:39,681 [INFO]: # total cells before filtering low-quality or volume outliers: 582360\n",
      "2025-09-17 09:54:39,681 [INFO]: # low_quality_cell:             5003\n",
      "2025-09-17 09:54:39,682 [INFO]: # volume_outlier_cell:       66\n",
      "2025-09-17 09:54:44,426 [INFO]: # total cells after filtering low-quality or volume outliers:  577291\n",
      "2025-09-17 09:54:44,427 [INFO]: # genes before filtering: 500\n",
      "2025-09-17 09:54:46,475 [INFO]: # genes after filtering: 500\n",
      "2025-09-17 09:54:48,020 [INFO]: Normalizing counts...\n",
      "2025-09-17 09:54:49,426 [INFO]: Cells before/after outlier removal during normalization: 577291 -> 565745\n",
      "/home/ubuntu/miniforge3/lib/python3.12/functools.py:909: UserWarning: zero-centering a sparse array/matrix densifies it.\n",
      "  return dispatch(args[0].__class__)(*args, **kw)\n"
     ]
    }
   ],
   "source": [
    "# Special case: rerun filter_low_quality_cells with lower threshold for vpt_3D\n",
    "if \"vpt_3D\" in args.seg_method:\n",
    "    logger.info(\n",
    "        \"Segmentation method contains 'vpt_3D': applying low-quality cell filtering with min_counts=10 due to smaller cell sizes.\"\n",
    "    )\n",
    "    adata = filter_low_quality_cells(\n",
    "        adata,\n",
    "        save_path=save_path / \"plots\",\n",
    "        min_counts=10,\n",
    "        logger=logger,\n",
    "    )\n",
    "else:\n",
    "    adata = filter_low_quality_cells(\n",
    "        adata,\n",
    "        save_path=save_path / \"plots\",\n",
    "        logger=logger,\n",
    "    )\n",
    "\n",
    "adata = filter_genes(adata, save_path=save_path / \"plots\", logger=logger)\n",
    "\n",
    "adata = normalize_counts(\n",
    "    adata, save_path=save_path / \"plots\", seg_method=args.seg_method, logger=logger\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6c620be2-37cb-434f-933d-3b0fe8d200d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-17 09:55:01,040 [INFO]: Dimensionality reduction: PCA\n",
      "2025-09-17 09:55:37,665 [INFO]: Dimensionality reduction: UMAP with {'n_neighbors': 20, 'n_pcs': 50, 'key': 'X_umap_20_50'}\n"
     ]
    }
   ],
   "source": [
    "adata = pca_umap_single(adata, save_path=save_path / \"plots\", logger=logger)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2dc9eb93-7279-4dcf-840b-18ef27be70ad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Looking for: ['bioconda::harmonypy']\n",
      "\n",
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Cache file \"/home/ubuntu/miniforge3/pkgs/cache/497deca9.json\" was modified by another program\n",
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Cache file \"/home/ubuntu/miniforge3/pkgs/cache/09cdf8bf.json\" was modified by another program\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.1s\n",
      "bioconda/linux-64 (check zst) \u001b[33m━━━━━━━━━━━━━━━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gbioconda/linux-64 (check zst)                       Checked  0.1s\n",
      "\u001b[?25h\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "bioconda/noarch (check zst) \u001b[33m━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gbioconda/noarch (check zst)                         Checked  0.0s\n",
      "\u001b[?25h\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "conda-forge/linux-64 \u001b[33m━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.1s\n",
      "conda-forge/linux-64 \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m  60.7kB /  47.0MB @   1.2MB/s  0.1s\n",
      "conda-forge/noarch   \u001b[33m━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.1s\n",
      "bioconda/linux-64    \u001b[33m━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.1s\n",
      "bioconda/noarch      \u001b[90m━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.2s\n",
      "conda-forge/linux-64 \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m   1.5MB /  47.0MB @   9.9MB/s  0.2s\n",
      "conda-forge/noarch   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m 205.3kB /  22.2MB @   1.3MB/s  0.2s\n",
      "bioconda/linux-64    \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m 188.2kB /   5.3MB @   1.2MB/s  0.2s\n",
      "bioconda/noarch      ━╸\u001b[90m━━━━━━━━━━━━━━━━━━━━━\u001b[0m 617.5kB /   5.0MB @   3.8MB/s  0.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.3s\n",
      "conda-forge/linux-64 ╸\u001b[90m━━━━━━━━━━━━━━━━━━━━━━\u001b[0m   4.2MB /  47.0MB @  15.9MB/s  0.3s\n",
      "conda-forge/noarch   ━╸\u001b[90m━━━━━━━━━━━━━━━━━━━━━\u001b[0m   2.8MB /  22.2MB @  10.5MB/s  0.3s\n",
      "bioconda/linux-64    ━━━━━━━━━━╸\u001b[90m━━━━━━━━━━━━\u001b[0m   2.8MB /   5.3MB @  10.3MB/s  0.3s\n",
      "bioconda/noarch      ━━━━━━━━━━━━━╸\u001b[90m━━━━━━━━━\u001b[0m   3.2MB /   5.0MB @  12.0MB/s  0.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gbioconda/noarch                                      5.0MB @  15.1MB/s  0.3s\n",
      "bioconda/linux-64                                    5.3MB @  15.0MB/s  0.4s\n",
      "[+] 0.4s\n",
      "conda-forge/linux-64 ━━╸\u001b[90m━━━━━━━━━━━━━━━━━━━━\u001b[0m   8.0MB /  47.0MB @  21.7MB/s  0.4s\n",
      "conda-forge/noarch   ━━━━━╸\u001b[90m━━━━━━━━━━━━━━━━━\u001b[0m   6.7MB /  22.2MB @  18.0MB/s  0.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.5s\n",
      "conda-forge/linux-64 ━━━━━╸\u001b[90m━━━━━━━━━━━━━━━━━\u001b[0m  14.2MB /  47.0MB @  30.0MB/s  0.5s\n",
      "conda-forge/noarch   ━━━━━━━━━━━━╸\u001b[90m━━━━━━━━━━\u001b[0m  12.9MB /  22.2MB @  27.1MB/s  0.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.6s\n",
      "conda-forge/linux-64 ━━━━━━━━╸\u001b[90m━━━━━━━━━━━━━━\u001b[0m  20.4MB /  47.0MB @  35.3MB/s  0.6s\n",
      "conda-forge/noarch   ━━━━━━━━━━━━━━━━━━╸\u001b[90m━━━━\u001b[0m  19.0MB /  22.2MB @  32.9MB/s  0.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gconda-forge/noarch                                  22.2MB @  35.1MB/s  0.7s\n",
      "[+] 0.7s\n",
      "conda-forge/linux-64 ━━━━━━━━━━━╸\u001b[90m━━━━━━━━━━━\u001b[0m  26.5MB /  47.0MB @  39.1MB/s  0.7s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.8s\n",
      "conda-forge/linux-64 ━━━━━━━━━━━━━━━━━╸\u001b[90m━━━━━\u001b[0m  37.7MB /  47.0MB @  48.3MB/s  0.8s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.9s\n",
      "conda-forge/linux-64 ━━━━━━━━━━━━━━━━━━━━━━  47.0MB @  54.4MB/s Downloaded  0.9s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gconda-forge/linux-64                               @  54.4MB/s  0.9s\n",
      "\u001b[?25h\n",
      "Pinned packages:\n",
      "  - python 3.12.*\n",
      "\n",
      "\n",
      "Transaction\n",
      "\n",
      "  Prefix: /home/ubuntu/miniforge3\n",
      "\n",
      "  Updating specs:\n",
      "\n",
      "   - bioconda::harmonypy\n",
      "   - ca-certificates\n",
      "   - certifi\n",
      "   - openssl\n",
      "\n",
      "\n",
      "  Package                Version  Build                 Channel           Size\n",
      "────────────────────────────────────────────────────────────────────────────────\n",
      "  Install:\n",
      "────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  \u001b[32m+ joblib         \u001b[0m        1.5.2  pyhd8ed1ab_0          conda-forge      225kB\n",
      "  \u001b[32m+ threadpoolctl  \u001b[0m        3.6.0  pyhecae5ae_0          conda-forge       24kB\n",
      "  \u001b[32m+ pytz           \u001b[0m       2025.2  pyhd8ed1ab_0          conda-forge      189kB\n",
      "  \u001b[32m+ python-tzdata  \u001b[0m       2025.2  pyhd8ed1ab_0          conda-forge      144kB\n",
      "  \u001b[32m+ six            \u001b[0m       1.17.0  pyhe01879c_1          conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ python-dateutil\u001b[0m  2.9.0.post0  pyhe01879c_2          conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libgfortran5   \u001b[0m       14.2.0  hf1ad2bd_2            conda-forge        1MB\n",
      "  \u001b[32m+ libgfortran    \u001b[0m       14.2.0  h69a702a_2            conda-forge       54kB\n",
      "  \u001b[32m+ libopenblas    \u001b[0m       0.3.29  pthreads_h94d23a6_0   conda-forge        6MB\n",
      "  \u001b[32m+ libblas        \u001b[0m        3.9.0  31_h59b9bed_openblas  conda-forge       17kB\n",
      "  \u001b[32m+ libcblas       \u001b[0m        3.9.0  31_he106b2a_openblas  conda-forge       17kB\n",
      "  \u001b[32m+ liblapack      \u001b[0m        3.9.0  31_h7ac8fdf_openblas  conda-forge       17kB\n",
      "  \u001b[32m+ numpy          \u001b[0m        2.3.3  py312h33ff503_0       conda-forge        9MB\n",
      "  \u001b[32m+ scipy          \u001b[0m       1.16.0  py312hf734454_0       conda-forge       17MB\n",
      "  \u001b[32m+ pandas         \u001b[0m        2.3.2  py312hf79963d_0       conda-forge       15MB\n",
      "  \u001b[32m+ scikit-learn   \u001b[0m        1.7.2  py312h4f0b9e3_0       conda-forge       10MB\n",
      "  \u001b[32m+ harmonypy      \u001b[0m       0.0.10  pyhdfd78af_0          bioconda          30kB\n",
      "\n",
      "  Upgrade:\n",
      "────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  \u001b[31m- ca-certificates\u001b[0m   2024.12.14  hbcca054_0            conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ ca-certificates\u001b[0m     2025.8.3  hbd8a1cb_0            conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- certifi        \u001b[0m   2024.12.14  pyhd8ed1ab_0          conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ certifi        \u001b[0m     2025.8.3  pyhd8ed1ab_0          conda-forge      159kB\n",
      "  \u001b[31m- openssl        \u001b[0m        3.4.0  h7b32b05_1            conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ openssl        \u001b[0m        3.5.3  h26f9b46_0            conda-forge        3MB\n",
      "\n",
      "  Summary:\n",
      "\n",
      "  Install: 17 packages\n",
      "  Upgrade: 3 packages\n",
      "\n",
      "  Total download: 62MB\n",
      "\n",
      "────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "Downloading  (2) \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m   0.0 B certifi                    0.0s\n",
      "Extracting       \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m       0                            0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.1s\n",
      "Downloading  (5) \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m  14.1kB certifi                    0.1s\n",
      "Extracting       \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m       0                            0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gthreadpoolctl                                       23.9kB @ 231.2kB/s  0.1s\n",
      "python-tzdata                                      144.2kB @   1.1MB/s  0.1s\n",
      "pytz                                               189.0kB @   1.2MB/s  0.2s\n",
      "joblib                                             224.7kB @   1.3MB/s  0.2s\n",
      "certifi                                            158.7kB @ 850.0kB/s  0.2s\n",
      "[+] 0.2s\n",
      "Downloading  (5) \u001b[33m━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━\u001b[0m 740.4kB libgfortran5               0.2s\n",
      "Extracting   (4) ╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━━━\u001b[0m       1 certifi                    0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gliblapack                                           16.8kB @  81.3kB/s  0.1s\n",
      "libgfortran5                                         1.5MB @   5.8MB/s  0.1s\n",
      "libcblas                                            16.8kB @  63.0kB/s  0.1s\n",
      "libopenblas                                          5.9MB @  20.5MB/s  0.2s\n",
      "[+] 0.3s\n",
      "Downloading  (5) ━━╸\u001b[33m━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━\u001b[0m  10.5MB harmonypy                  0.3s\n",
      "Extracting   (3) ━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━\u001b[0m       6 libcblas                   0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gopenssl                                              3.1MB @   9.8MB/s  0.1s\n",
      "libblas                                             16.9kB @  50.0kB/s  0.1s\n",
      "harmonypy                                           30.0kB @  88.9kB/s  0.0s\n",
      "libgfortran                                         53.7kB @ 136.9kB/s  0.1s\n",
      "[+] 0.4s\n",
      "Downloading  (4) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m  25.5MB numpy                      0.4s\n",
      "Extracting   (5) ━━━━━━━━━╸\u001b[33m━━━━━━━╸\u001b[0m\u001b[90m━━━━━\u001b[0m       8 harmonypy                  0.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gscikit-learn                                         9.7MB @  22.7MB/s  0.2s\n",
      "pandas                                              15.1MB @  34.9MB/s  0.3s\n",
      "numpy                                                8.8MB @  18.7MB/s  0.1s\n",
      "[+] 0.5s\n",
      "Downloading  (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m  54.9MB scipy                      0.5s\n",
      "Extracting   (3) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━\u001b[0m      13 numpy                      0.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gscipy                                               16.8MB @  31.1MB/s  0.2s\n",
      "[+] 0.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━  61.8MB                            0.6s\n",
      "Extracting   (4) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      13 numpy                      0.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━  61.8MB                            0.6s\n",
      "Extracting   (4) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      13 numpy                      0.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━  61.8MB                            0.6s\n",
      "Extracting   (2) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m      15 pandas                     0.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━  61.8MB                            0.6s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━\u001b[0m      16 pandas                     0.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G\u001b[?25h\n",
      "Downloading and Extracting Packages:\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "!mamba install bioconda::harmonypy -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e7eef546-47b1-4c61-b504-861269a92ad8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-17 10:09:10,435 [INFO]: Integration: Run Harmony\n",
      "2025-09-17 10:09:10,853 - harmonypy - INFO - Computing initial centroids with sklearn.KMeans...\n",
      "2025-09-17 10:09:57,750 - harmonypy - INFO - sklearn.KMeans initialization complete.\n",
      "2025-09-17 10:09:59,675 - harmonypy - INFO - Iteration 1 of 10\n",
      "2025-09-17 10:12:56,241 - harmonypy - INFO - Iteration 2 of 10\n",
      "2025-09-17 10:15:53,019 - harmonypy - INFO - Iteration 3 of 10\n",
      "2025-09-17 10:18:49,596 - harmonypy - INFO - Iteration 4 of 10\n",
      "2025-09-17 10:21:37,827 - harmonypy - INFO - Iteration 5 of 10\n",
      "2025-09-17 10:22:31,845 - harmonypy - INFO - Iteration 6 of 10\n",
      "2025-09-17 10:23:25,674 - harmonypy - INFO - Iteration 7 of 10\n",
      "2025-09-17 10:24:19,556 - harmonypy - INFO - Iteration 8 of 10\n",
      "2025-09-17 10:25:13,462 - harmonypy - INFO - Converged after 8 iterations\n",
      "2025-09-17 10:25:13,527 [INFO]: Integration: Compute neighbors and UMAP\n"
     ]
    }
   ],
   "source": [
    "adata = integration_harmony(\n",
    "    adata,\n",
    "    batch_key=\"slide\",\n",
    "    save_path=save_path / \"plots\",\n",
    "    logger=logger,  ################\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aab903b-78b2-4a0b-bbe4-c799beb4fb67",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "98df782b-9457-473c-ae0b-6d37afe1bc33",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "from os.path import join\n",
    "from typing import Tuple\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "from anndata import AnnData\n",
    "from matplotlib.path import Path\n",
    "\n",
    "\n",
    "def _plot_integration_comparison(\n",
    "    adata: AnnData,\n",
    "    save_path: str,\n",
    "    umap_key: str,\n",
    "    batch_key: str,\n",
    "    point_size_factor: int = 320000,\n",
    ") -> None:\n",
    "    \"\"\"Helper function to plot before/after integration comparison.\"\"\"\n",
    "\n",
    "    fig, axes = plt.subplots(\n",
    "        4, 2, figsize=(17, 22), gridspec_kw={\"hspace\": 0.01, \"wspace\": -0.54}\n",
    "    )\n",
    "\n",
    "    fig.text(0.40, 0.89, \"Unintegrated\", fontsize=16, ha=\"center\")\n",
    "    fig.text(\n",
    "        0.62,\n",
    "        0.89,\n",
    "        f\"Integrated (Harmony{f'; by {batch_key}' if batch_key else ''})\",\n",
    "        fontsize=16,\n",
    "        ha=\"center\",\n",
    "    )\n",
    "\n",
    "    plot_configs = [\n",
    "        (\"sample\", \"Sample\"),\n",
    "        (\"slide\", \"Slide\"),\n",
    "        (\"condition\", \"Condition\"),\n",
    "        (\"cell_type_mmc_raw_revised\", \"Cell Type\"),\n",
    "    ]\n",
    "\n",
    "    for i, (color_key, label) in enumerate(plot_configs):\n",
    "        # Unintegrated\n",
    "        sc.pl.embedding(\n",
    "            adata,\n",
    "            basis=umap_key.replace(\"_harmony\", \"\"),\n",
    "            color=color_key,\n",
    "            show=False,\n",
    "            ax=axes[i, 0],\n",
    "            size=point_size_factor / adata.shape[0],\n",
    "            legend_loc=None,\n",
    "            title=\"\",\n",
    "        )\n",
    "\n",
    "        # Integrated\n",
    "        sc.pl.embedding(\n",
    "            adata,\n",
    "            basis=umap_key,\n",
    "            color=color_key,\n",
    "            show=False,\n",
    "            ax=axes[i, 1],\n",
    "            size=point_size_factor / adata.shape[0],\n",
    "            legend_loc=\"right margin\",\n",
    "            title=\"\",\n",
    "        )\n",
    "\n",
    "        axes[i, 0].text(\n",
    "            -0.05,\n",
    "            0.5,\n",
    "            label,\n",
    "            transform=axes[i, 0].transAxes,\n",
    "            fontsize=14,\n",
    "            rotation=90,\n",
    "            va=\"center\",\n",
    "            ha=\"center\",\n",
    "        )\n",
    "\n",
    "    for ax in axes.ravel():\n",
    "        ax.set_aspect(\"equal\", adjustable=\"box\")\n",
    "        ax.set_xlabel(\"\")\n",
    "        ax.set_ylabel(\"\")\n",
    "\n",
    "    plt.savefig(\n",
    "        join(save_path, \"UMAP_integrated_harmony2.png\"), dpi=150, bbox_inches=\"tight\"\n",
    "    )\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "88f1cb9d-0047-4cd0-adba-10c9ce6a9313",
   "metadata": {},
   "outputs": [],
   "source": [
    "_plot_integration_comparison(\n",
    "    adata,\n",
    "    save_path=save_path / \"plots\",\n",
    "    umap_key=\"X_umap_harmony_20_50\",\n",
    "    batch_key=\"slide\",\n",
    "    point_size_factor=200000,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72be87be-343b-4e1f-96e7-f32bfa8ce123",
   "metadata": {},
   "source": [
    "# tbc here\n",
    "- add plotting code above to script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ab0f3c82-d5de-48a1-8169-51e476839ee3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-17 10:58:27,720 [INFO]: Saving integrated object...\n",
      "2025-09-17 11:00:46,394 [INFO]: Done.\n"
     ]
    }
   ],
   "source": [
    "# Save result\n",
    "logger.info(\"Saving integrated object...\")\n",
    "if \"fov\" not in adata.obs.columns:\n",
    "    adata.obs[\"fov\"] = \"\"\n",
    "adata.obs[\"fov\"] = adata.obs[\"fov\"].astype(str)\n",
    "output_path = save_path / \"adatas\"\n",
    "output_path.mkdir(parents=True, exist_ok=True)\n",
    "adata.write(output_path / \"adata_integrated.h5ad.gz\", compression=\"gzip\")\n",
    "logger.info(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf2a5ddd-86b8-4c69-94b9-f12cbf1087dd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33b42cfd-09b6-4af2-be9a-4cff9164842a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c5a84ea8-6816-4faa-bf7c-cf6d047e5708",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-04 11:44:15,699 [INFO]: Loading data...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 35.6 s, sys: 4.15 s, total: 39.8 s\n",
      "Wall time: 42.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "logger.info(\"Loading data...\")\n",
    "adata = sc.read_h5ad(save_path / \"adatas\" / \"adata_integrated.h5ad.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1a818db4-8cdc-40e1-9a87-522e61741006",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/dss/dssfs03/pn52re/pn52re-dss-0001/cellseg-benchmark/analysis/foxf2/Cellpose_1_nuclei_model')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "save_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "da1d9d78-b95a-450a-8e36-4fdeb254c4a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.drop(columns=[\"spt_region\"], inplace=True, errors=\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e99998-6424-45dd-9f5c-4fe098a0b13f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
